## Loops to read many files

Let's work through an analysis of the size of "structural" viral proteins.
We've done some of the work for you already, and would recommend that you
always document where the data came from and how it was retrieved/processed:

> The NCBI Identital Protein Groups database was queried for "structural"
> with "Division" restricted to "Viruses" and "Source database" restricted
> to "UniProtKB/Swiss-Prot". 243 results were retrieved as FASTA files and 
> converted using BASH commands into separate tab-delimited files.
^[One crucial aspect missing is the actual BASH commands used. I did not
    save them as a script, and I *should* go back, save them in a script,
    and re-run that to make sure I get the same result!]

These files are in the `data/viral_structural_proteins` folder,
are tab-delimited three fields, and look like this:

    Q8V433.1 Membrane protein       Bovine respiratory coronavirus (strain 98TXSF-110-LUN)  MSSVTTPAPVYTWTA...

How do you inspect these files?
What BASH tools do you use?

---

### Reading in one file

Let's read one of them into R.

The records are "delimited" by tabs, so tab-separated.
We'll need to use `read.delim`, specify a tab separator, and ignore the lack
of a header. How do we do this? 

#### Review of looking up documentation

You can look up the documentation for any named function or package by using
the `?function` syntax.

Sometimes, you need to use extra backticks to make it work, like "?`+`"

If you don't know what you're looking for, you can search with `??`.
Extra backticks don't hurt, and help when you have spaces in the query.

---

#### Back to the file-reading

What does `?read.delim` say to do?

```{=html}
<div class="incremental">
```


```{r,calculate_one,cache=T}
viral_protein_data <- read.delim("data/viral_structural_proteins/viral_proteins_100.tsv",
    sep="\t",header=F)
viral_protein_data
```

- How do we find the column names?
- What are two ways to access the protein sequence?
- How do we calculate the number of characters in this sequence?
    Try searching with "??`number of characters`"

```{=html}
</div>
```

---


### Doing things multiple times


You will want to do the same thing multiple
times - automation is great. Let's code up a
simple little analysis - we'll calculate the
length (number of amino acids) in a viral
structural protein sequence.

Let's do this for all the proteins.
The simplest way of doing this is to copy and
paste it.

```{r,copy_paste_rep,cache=T}
viral_protein_data <- read.delim("data/viral_structural_proteins/viral_proteins_002.tsv",
      sep="\t",header=F, as.is=T)
nchar(viral_protein_data$V3[[1]])
viral_protein_data <- read.delim("data/viral_structural_proteins/viral_proteins_003.tsv",
      sep="\t",header=F, as.is=T)
nchar(viral_protein_data[1,"V3"])
```

Go ahead and do this for all 242 proteins 

---

... just kidding.

Try this to, well, list files:

```{r,list_files,cache=T}
list.files(path="data/viral_structural_proteins")[1:5]
```

Note that I put a `[1:5]` to limit it to the first 5.
You could also use `head()`. It's a good idea to work
with a small subset of files while you are iterating
through development, then scale it up to the entirety.

Now how do you calculate the number of characters for each protein?

How do we do this for every file listed ... ?

---

### Loops

<!-- video -->

Loops are for running a "code block" as many times
as the "condition" determines.

- A "code block" is either one line of code, or multiple
    lines of code surrounded by curly brackets - {}

        {
            code <- "in a block" 
            with <- "multiple lines" 
        }

- A "condition" is an expression of code that
    can either evaluate to either `TRUE` or `FALSE`, or 
    set a variable for each time running the code block.

The most common form of this is a `for` loop.
Other "control statements" or "flow control statements" are
`while`, `repeat`, and `if`.

Let's look up what they do, with "?`for`"
Note the back ticks! These are a trick in R to make anything
be interpreted as literally what you type, and not
any special characters. Like "?`+`", or "?`?`"

---

Here's an example loop:

```{r,loop_example_1,cache=T}
for (i in 1:4) {
  print(i)
}
```

The pieces:

- `(i in 1:4)` is what is being looped over - `1:4` is a vector of 
  1 through 4 that is created, and it is put one at a time into `i` 
  (a new variable). You need the parentheses.
  
- `{` and `}` denote the opening and closing brackets, specify the
  "code block" that is run each time.
  
- inside this "code block" is `print(i)` - it prints the variable `i`,
  which is set to a value of 1, 2, 3, or 4 for each loop

---

How do we loop through and print each file name?

What are the (1) code block and (2) loop condition ?

```{=html}
<div class="incremental">
```


```{r,loop_through_files,cache=T}
for (i in list.files(path="data/viral_structural_proteins")[1:5] ) {
  print(i)
}
```

How do we modify this to calculate the protein length?

```{r,loop_and_calculate_errors,cache=T,error=T}
for (i in list.files(path="data/viral_structural_proteins")[1:5] ) {
  print(i)
  print(nchar(read.delim(i)$V3[[1]]))
}
```

```{=html}
</div>
```

---

Debugging...

```{r,loop_and_calculate,cache=T}
for (i in list.files(path="data/viral_structural_proteins",
                     full.names=T)[1:5] ) {
  print(nchar(read.delim(i,sep="\t",header=F,as.is=T)$V3[[1]]))
}
```

---

#### Storing values from a loop

How do we store this value?

In other languages, "append". But R is not built that way. It'll work,
but it's very inefficient. The "R-way" to store values from a loop
is to define a vector of the right length, 
then put each element in it.

```{r,vector_of_size,cache=T}
vector("character",10)
vector(mode="numeric",length=5)
vector("logical",2)
```

---

Okay, but how do we access each position to save the value?
We need to turn out list of files into indicies.
We'll save it first so we can count how many there are.

`seq_along` is handy function, otherwise use something like
`seq(1,length(x))`.

```{r,loop_save_and_index,cache=T}
first_five <- list.files(path="data/viral_structural_proteins")[1:5]
for (i in seq_along(first_five) ) {
  print(first_five[i])
}
```

---

Putting these together, we can save a vector of file names:

```{r,loop_save_and_index_filenames,cache=T}
first_five <- list.files(path="data/viral_structural_proteins")[1:5]
filenamez <- vector("character",length(first_five))

for (i in seq_along(first_five) ) {
  filenamez[i] <- first_five[i]
}

filenamez
```

---

And finally calculate the length of each protein:

```{r,loop_to_calc_lengths_work,cache=T,error=T}
first_five <- list.files(path="data/viral_structural_proteins",
                         full.names=T, pattern=".*tsv")[1:5]
lengthz <- vector("character",length(first_five))

for (i in seq_along(first_five) ) {
  lengthz[i] <- nchar( read.delim(first_five[i],
      header=F,as.is=T,sep="\t")$V3[[1]]
    )
}

lengthz
```

---

Now we can take off the `[1:5]` limiter, and do the whole set:

```{r,loop_to_calc_lengths_all,cache=T,error=T}
first_five <- list.files(path="data/viral_structural_proteins",
                         full.names=T, pattern=".*tsv")
lengthz <- vector("character",length(first_five))

for (i in seq_along(first_five) ) {
  lengthz[i] <- nchar( read.delim(first_five[i],
      header=F,as.is=T,sep="\t")$V3[[1]]
    )
}
```

---

How do we go about visualizing/analyzing this?

```{r,base_viz_protein_lengths_error,cache=T,error=T}
hist(lengthz)
```

er what...?

Numeric....

```{r,loop_to_calc_lengths_all_fixed,cache=T,error=T}
first_five <- list.files(path="data/viral_structural_proteins",
                         full.names=T, pattern=".*tsv")
lengthz <- vector("numeric",length(first_five))

for (i in seq_along(first_five) ) {
  lengthz[i] <- nchar( read.delim(first_five[i],
      header=F,as.is=T,sep="\t")$V3[[1]]
    )
}
```

---

```{r,base_viz_protein_lengths_fix,cache=T,error=T}
hist(lengthz,breaks=50)
```

---

```{r,ggplot_viz_protein_lengths,cache=T,error=T}
library(ggplot2)
ggplot( data.frame(length=lengthz) )+theme_classic()+
  aes(x=length)+geom_histogram(bins=50)

summary(lengthz)
```

---



