<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.2 Analyzing viral protein sequences | DLL 2021, R section</title>
  <meta name="description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="4.2 Analyzing viral protein sequences | DLL 2021, R section" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="github-repo" content="darachm/dll-r" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.2 Analyzing viral protein sequences | DLL 2021, R section" />
  
  <meta name="twitter:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  

<meta name="author" content="mesako" />
<meta name="author" content="Margaret" />
<meta name="author" content="Samson" />
<meta name="author" content="Zac" />
<meta name="author" content="darachm" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="what-is-a-workflow.html"/>
<link rel="next" href="packages-install-and-use.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DLL 2021 R workshops</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Workshop Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="workshop-goals.html"><a href="workshop-goals.html"><i class="fa fa-check"></i><b>1.1</b> Workshop goals</a></li>
<li class="chapter" data-level="1.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html"><i class="fa fa-check"></i><b>1.2</b> Structure and resources</a><ul>
<li class="chapter" data-level="1.2.1" data-path="structure-and-resources.html"><a href="structure-and-resources.html#workshop-schedule"><i class="fa fa-check"></i><b>1.2.1</b> Workshop schedule</a></li>
<li class="chapter" data-level="1.2.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html#asynchronous-sessions"><i class="fa fa-check"></i><b>1.2.2</b> Asynchronous sessions</a></li>
<li class="chapter" data-level="1.2.3" data-path="structure-and-resources.html"><a href="structure-and-resources.html#synchronous-sessions"><i class="fa fa-check"></i><b>1.2.3</b> Synchronous sessions</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="workshop-expectations.html"><a href="workshop-expectations.html"><i class="fa fa-check"></i><b>1.3</b> Workshop expectations</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="day-3-intoduction-to-r.html"><a href="day-3-intoduction-to-r.html"><i class="fa fa-check"></i><b>2</b> Day 3 - intoduction to R</a><ul>
<li class="chapter" data-level="2.1" data-path="using-rstudio.html"><a href="using-rstudio.html"><i class="fa fa-check"></i><b>2.1</b> Using RStudio</a><ul>
<li class="chapter" data-level="2.1.1" data-path="using-rstudio.html"><a href="using-rstudio.html#r-scripts-in-rstudio"><i class="fa fa-check"></i><b>2.1.1</b> R Scripts in RStudio</a></li>
<li class="chapter" data-level="2.1.2" data-path="using-rstudio.html"><a href="using-rstudio.html#console-and-terminal"><i class="fa fa-check"></i><b>2.1.2</b> Console and Terminal</a></li>
<li class="chapter" data-level="2.1.3" data-path="using-rstudio.html"><a href="using-rstudio.html#environment-and-history"><i class="fa fa-check"></i><b>2.1.3</b> Environment and History</a></li>
<li class="chapter" data-level="2.1.4" data-path="using-rstudio.html"><a href="using-rstudio.html#files-plots-and-packages"><i class="fa fa-check"></i><b>2.1.4</b> Files, Plots, and Packages</a></li>
<li class="chapter" data-level="2.1.5" data-path="using-rstudio.html"><a href="using-rstudio.html#putting-it-all-together"><i class="fa fa-check"></i><b>2.1.5</b> Putting It All Together</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-rmds.html"><a href="using-rmds.html"><i class="fa fa-check"></i><b>2.2</b> Using Rmds</a><ul>
<li class="chapter" data-level="2.2.1" data-path="using-rmds.html"><a href="using-rmds.html#basic-structure"><i class="fa fa-check"></i><b>2.2.1</b> Basic Structure</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-rmds.html"><a href="using-rmds.html#overview-of-markdown"><i class="fa fa-check"></i><b>2.2.2</b> Overview of Markdown</a></li>
<li class="chapter" data-level="2.2.3" data-path="using-rmds.html"><a href="using-rmds.html#overview-of-how-to-write-an-r-chunk"><i class="fa fa-check"></i><b>2.2.3</b> Overview of how to write an R chunk</a></li>
<li class="chapter" data-level="2.2.4" data-path="using-rmds.html"><a href="using-rmds.html#final-thoughts"><i class="fa fa-check"></i><b>2.2.4</b> Final Thoughts</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Getting started in R</a><ul>
<li class="chapter" data-level="2.3.1" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#getting-help-in-r"><i class="fa fa-check"></i><b>2.3.1</b> Getting help in R</a></li>
<li class="chapter" data-level="2.3.2" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-variables-and-data-types"><i class="fa fa-check"></i><b>2.3.2</b> Using variables and data types</a></li>
<li class="chapter" data-level="2.3.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-functions"><i class="fa fa-check"></i><b>2.3.3</b> Using functions</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html"><i class="fa fa-check"></i><b>2.4</b> Working with vectors</a><ul>
<li class="chapter" data-level="2.4.1" data-path="working-with-vectors.html"><a href="working-with-vectors.html#building-vectors"><i class="fa fa-check"></i><b>2.4.1</b> Building vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="working-with-vectors.html"><a href="working-with-vectors.html#indexing-and-subsetting-vectors"><i class="fa fa-check"></i><b>2.4.2</b> Indexing and subsetting vectors</a></li>
<li class="chapter" data-level="2.4.3" data-path="working-with-vectors.html"><a href="working-with-vectors.html#using-logic-and-logicals"><i class="fa fa-check"></i><b>2.4.3</b> Using logic and logicals</a></li>
<li class="chapter" data-level="2.4.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html#modifying-vectors"><i class="fa fa-check"></i><b>2.4.4</b> Modifying vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html"><i class="fa fa-check"></i><b>2.5</b> Working with dataframes</a><ul>
<li class="chapter" data-level="2.5.1" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#indexing-and-subsetting-dataframes"><i class="fa fa-check"></i><b>2.5.1</b> Indexing and subsetting dataframes</a></li>
<li class="chapter" data-level="2.5.2" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#exploring-dataframes"><i class="fa fa-check"></i><b>2.5.2</b> Exploring dataframes</a></li>
<li class="chapter" data-level="2.5.3" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#building-and-modifying-dataframes"><i class="fa fa-check"></i><b>2.5.3</b> Building and modifying dataframes</a></li>
<li class="chapter" data-level="2.5.4" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#handling-datatypes-in-dataframes"><i class="fa fa-check"></i><b>2.5.4</b> Handling datatypes in dataframes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="day-4-tidyverse-and-visualizations.html"><a href="day-4-tidyverse-and-visualizations.html"><i class="fa fa-check"></i><b>3</b> Day 4 - tidyverse and visualizations</a><ul>
<li class="chapter" data-level="3.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html"><i class="fa fa-check"></i><b>3.1</b> Reading and processing data</a><ul>
<li class="chapter" data-level="3.1.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#loading-a-file-into-r"><i class="fa fa-check"></i><b>3.1.1</b> Loading a file into R</a></li>
<li class="chapter" data-level="3.1.2" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#handling-issues-in-data"><i class="fa fa-check"></i><b>3.1.2</b> Handling issues in data</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html"><i class="fa fa-check"></i><b>3.2</b> Tidyverse 1/2: functions &amp; pipes</a><ul>
<li class="chapter" data-level="3.2.1" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#welcome-to-the-tidyverse"><i class="fa fa-check"></i><b>3.2.1</b> Welcome to the tidyverse!</a></li>
<li class="chapter" data-level="3.2.2" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#load-tidyverse"><i class="fa fa-check"></i><b>3.2.2</b> Load tidyverse</a></li>
<li class="chapter" data-level="3.2.3" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#how-to-pipe"><i class="fa fa-check"></i><b>3.2.3</b> How to pipe %&gt;%</a></li>
<li class="chapter" data-level="3.2.4" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#essential-tidyverse-functions"><i class="fa fa-check"></i><b>3.2.4</b> Essential tidyverse functions</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html"><i class="fa fa-check"></i><b>3.3</b> Tidyverse 2/2: reshaping data</a><ul>
<li class="chapter" data-level="3.3.1" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html#why-tidy-data"><i class="fa fa-check"></i><b>3.3.1</b> Why tidy data</a></li>
<li class="chapter" data-level="3.3.2" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html#how-to-tidy-data"><i class="fa fa-check"></i><b>3.3.2</b> How to tidy data</a></li>
<li class="chapter" data-level="3.3.3" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html#data-exploration-with-tidyverse"><i class="fa fa-check"></i><b>3.3.3</b> Data exploration with tidyverse</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html"><i class="fa fa-check"></i><b>3.4</b> Making plots with ggplot2</a><ul>
<li class="chapter" data-level="3.4.1" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#getting-started-with-a-ggplot"><i class="fa fa-check"></i><b>3.4.1</b> Getting started with a ggplot</a></li>
<li class="chapter" data-level="3.4.2" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#changing-plot-aesthetics"><i class="fa fa-check"></i><b>3.4.2</b> Changing plot aesthetics</a></li>
<li class="chapter" data-level="3.4.3" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#exploring-simple-plots"><i class="fa fa-check"></i><b>3.4.3</b> Exploring simple plots</a></li>
<li class="chapter" data-level="3.4.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#visualizing-between-groups"><i class="fa fa-check"></i><b>3.4.4</b> Visualizing between groups</a></li>
<li class="chapter" data-level="3.4.5" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#generating-heatmaps"><i class="fa fa-check"></i><b>3.4.5</b> Generating heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html"><i class="fa fa-check"></i><b>3.5</b> Making scientific figures</a><ul>
<li class="chapter" data-level="3.5.1" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#plotting-error-bars"><i class="fa fa-check"></i><b>3.5.1</b> Plotting error bars</a></li>
<li class="chapter" data-level="3.5.2" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#showing-trends-in-data"><i class="fa fa-check"></i><b>3.5.2</b> Showing trends in data</a></li>
<li class="chapter" data-level="3.5.3" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#saving-figures-locally"><i class="fa fa-check"></i><b>3.5.3</b> Saving figures locally</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html"><i class="fa fa-check"></i><b>3.6</b> Applying basic stats</a><ul>
<li class="chapter" data-level="3.6.1" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#describing-quantitative-variables"><i class="fa fa-check"></i><b>3.6.1</b> Describing quantitative variables</a></li>
<li class="chapter" data-level="3.6.2" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#finding-differences-in-data"><i class="fa fa-check"></i><b>3.6.2</b> Finding differences in data</a></li>
<li class="chapter" data-level="3.6.3" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#identifying-correlations"><i class="fa fa-check"></i><b>3.6.3</b> Identifying correlations</a></li>
<li class="chapter" data-level="3.6.4" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#producing-linear-models"><i class="fa fa-check"></i><b>3.6.4</b> Producing linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="day-5-workflows-loops-packages-and-help.html"><a href="day-5-workflows-loops-packages-and-help.html"><i class="fa fa-check"></i><b>4</b> Day 5 - workflows, loops, packages, and help</a><ul>
<li class="chapter" data-level="4.1" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html"><i class="fa fa-check"></i><b>4.1</b> What is a workflow?</a><ul>
<li class="chapter" data-level="4.1.1" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html#organizing-files"><i class="fa fa-check"></i><b>4.1.1</b> Organizing files</a></li>
<li class="chapter" data-level="4.1.2" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html#use-a-notebook-to-generate-reports"><i class="fa fa-check"></i><b>4.1.2</b> Use a notebook to generate reports</a></li>
<li class="chapter" data-level="4.1.3" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html#sharing-with-others"><i class="fa fa-check"></i><b>4.1.3</b> Sharing with others</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html"><i class="fa fa-check"></i><b>4.2</b> Analyzing viral protein sequences</a><ul>
<li class="chapter" data-level="4.2.1" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#reading-in-tab-delimited-file"><i class="fa fa-check"></i><b>4.2.1</b> Reading in tab-delimited file</a></li>
<li class="chapter" data-level="4.2.2" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#string-stuff"><i class="fa fa-check"></i><b>4.2.2</b> String stuff</a></li>
<li class="chapter" data-level="4.2.3" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#loops---doing-a-similar-task-multiple-times"><i class="fa fa-check"></i><b>4.2.3</b> Loops - doing a similar task multiple times</a></li>
<li class="chapter" data-level="4.2.4" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#storing-values-from-a-loop"><i class="fa fa-check"></i><b>4.2.4</b> Storing values from a loop</a></li>
<li class="chapter" data-level="4.2.5" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#extending-the-workflow"><i class="fa fa-check"></i><b>4.2.5</b> Extending the workflow</a></li>
<li class="chapter" data-level="4.2.6" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#conclusion"><i class="fa fa-check"></i><b>4.2.6</b> Conclusion</a></li>
<li class="chapter" data-level="4.2.7" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#totally-optional-supplements-for-your-edification"><i class="fa fa-check"></i><b>4.2.7</b> Totally optional supplements, for your edification</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html"><i class="fa fa-check"></i><b>4.3</b> Packages - install and use</a><ul>
<li class="chapter" data-level="4.3.1" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#cran-is-the-canonical-r-package-resource"><i class="fa fa-check"></i><b>4.3.1</b> CRAN is the canonical R package resource</a></li>
<li class="chapter" data-level="4.3.2" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#using-packages-without-library-ing-them"><i class="fa fa-check"></i><b>4.3.2</b> Using packages without <code>library()</code>-ing them</a></li>
<li class="chapter" data-level="4.3.3" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#github-is-a-common-place-for-sharing-packges-in-development"><i class="fa fa-check"></i><b>4.3.3</b> Github is a common place for sharing packges in development</a></li>
<li class="chapter" data-level="4.3.4" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#bioconductor-is-a-great-place-for-bio-related-packages"><i class="fa fa-check"></i><b>4.3.4</b> Bioconductor is a great place for bio-related packages</a></li>
<li class="chapter" data-level="4.3.5" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#what-packages-do-you-have-already"><i class="fa fa-check"></i><b>4.3.5</b> What packages do you have already?</a></li>
<li class="chapter" data-level="4.3.6" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#major-problems-with-package-installations"><i class="fa fa-check"></i><b>4.3.6</b> Major problems with package installations</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html"><i class="fa fa-check"></i><b>4.4</b> Getting help - troubleshooting</a><ul>
<li class="chapter" data-level="4.4.1" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#where-to-start-general-help"><i class="fa fa-check"></i><b>4.4.1</b> Where to start? General help</a></li>
<li class="chapter" data-level="4.4.2" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#debugging-specific-problems"><i class="fa fa-check"></i><b>4.4.2</b> Debugging specific problems</a></li>
<li class="chapter" data-level="4.4.3" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#minimal-reproducible-examples"><i class="fa fa-check"></i><b>4.4.3</b> Minimal reproducible examples</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="the-struggle.html"><a href="the-struggle.html"><i class="fa fa-check"></i><b>4.5</b> The Struggle</a></li>
<li class="chapter" data-level="4.6" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>4.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="appendix-links.html"><a href="appendix-links.html"><i class="fa fa-check"></i><b>5</b> Appendix - links</a><ul>
<li class="chapter" data-level="5.1" data-path="ideas-for-figures.html"><a href="ideas-for-figures.html"><i class="fa fa-check"></i><b>5.1</b> Ideas for figures</a></li>
<li class="chapter" data-level="5.2" data-path="more-links.html"><a href="more-links.html"><i class="fa fa-check"></i><b>5.2</b> More links</a></li>
<li class="chapter" data-level="5.3" data-path="more-about-workflows.html"><a href="more-about-workflows.html"><i class="fa fa-check"></i><b>5.3</b> More about workflows</a></li>
<li class="chapter" data-level="5.4" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html"><i class="fa fa-check"></i><b>5.4</b> Style - writing (re)usable code</a><ul>
<li class="chapter" data-level="5.4.1" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#functions-as-modular-steps"><i class="fa fa-check"></i><b>5.4.1</b> Functions as modular steps</a></li>
<li class="chapter" data-level="5.4.2" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#tips-for-modular-workflows"><i class="fa fa-check"></i><b>5.4.2</b> Tips for modular workflows</a></li>
<li class="chapter" data-level="5.4.3" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#apply-is-a-popular-tool"><i class="fa fa-check"></i><b>5.4.3</b> Apply is a popular tool</a></li>
<li class="chapter" data-level="5.4.4" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#learn-from-examples"><i class="fa fa-check"></i><b>5.4.4</b> Learn from examples</a></li>
<li class="chapter" data-level="5.4.5" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#style-guides-can-be-inspiring"><i class="fa fa-check"></i><b>5.4.5</b> Style guides can be inspiring</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DLL 2021, R section</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="analyzing-viral-protein-sequences" class="section level2">
<h2><span class="header-section-number">4.2</span> Analyzing viral protein sequences</h2>
<p>Let’s practice with starting a project.</p>
<p>Viral structural proteins are dramatically more important to the general
public than they were 2 years ago - and under much more public attention.
Let’s take a cursory look at some of these.</p>
<p>What kind of data are we using for this?
We’ll retrieve some determined protein sequences from NCBI databases.
We’ve already done that work for you already. Here’s a short writeup of
what we did, amd we would recommend that you
always document where the data came from and how it was retrieved/processed:</p>
<blockquote>
<p>The NCBI Identical Protein Groups database was queried for “structural”
with “Division” restricted to “Viruses” and “Source database” restricted
to “UniProtKB/Swiss-Prot”. 243 results were retrieved as FASTA files and
converted using bash commands into separate tab-delimited files.
<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
</blockquote>
<p>What format are these in?
These files are in the <code>data/viral_structural_proteins</code> folder,
they end in <code>.tsv</code>, are tab-delimited three fields, and look like this:</p>
<pre><code>Q8V433.1 Membrane protein       Bovine respiratory coronavirus (strain 98TXSF-110-LUN)  MSSVTTPAPVYTWTA...</code></pre>
<p>This is not a standard format, because I’d like to tell y’all about
reading tab-delimited files.
<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a></p>
<p>One trick to look at files using R is using <code>readLines()</code> function.
This just reads lines, up to the argument <code>n=</code> number of lines.
Try something like:</p>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb337-1" title="1"><span class="kw">readLines</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_242.tsv&quot;</span>,<span class="dt">n=</span><span class="dv">5</span>)</a></code></pre></div>
<pre><code>## [1] &quot;YP_009724393.1 membrane glycoprotein \tSevere acute respiratory syndrome coronavirus 2\tMADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFLYIIKLIFLWLLWPVTLACFVLAAVYRINWITGGIAIAMACLVGLMWLSYFIASFRLFARTRSMWSFNPETNILLNVPLHGTILTRPLLESELVIGAVILRGHLRIAGHHLGRCDIKDLPKEITVATSRTLSYYKLGASQRVAGDSGFAAYSRYRIGNYKLNTDHSSSSDNIALLVQ&quot;</code></pre>
<p>Whenever using these commands, put your cursor in the quotes and then hit TAB,
to use autocomplete.</p>
<p>Let’s do two things. Let’s count how long the protein sequences are, so
we can look at that distribution. Let’s also look for particular combinations
of amino acids.</p>
<p>Some folks are quite interested in the insertion of an amino-acid sequence
of <code>PRRA</code>, relative to other related coronaviruses, in the SARS-CoV-2 genome.
I don’t know hardly anything about furin-cleavage sites or viral evolution,
but since this is an introductory R class,
we can use the tools we have to look for this sequence in the files.
Let’s do that<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a></p>
<p>How?</p>
<hr />
<div id="reading-in-tab-delimited-file" class="section level3">
<h3><span class="header-section-number">4.2.1</span> Reading in tab-delimited file</h3>
<p>Let’s read one of them into R.</p>
<p>The records are “delimited” by tabs, so each field is tab-separated.
We’ll need to use the <code>read.delim</code> function and specify a tab separator.</p>
<hr />
<div id="review-of-looking-up-documentation" class="section level4">
<h4><span class="header-section-number">4.2.1.1</span> Review of looking up documentation</h4>
<p>You can look up the documentation for any named function or package by using
the <code>?function</code> syntax.</p>
<p>Sometimes, you need to use extra backticks to make it work, like <code>?`+`</code></p>
<p>If you don’t know what you’re looking for, you can search with <code>??</code>.
Extra backticks don’t hurt, and are necessary when you have spaces in the
query.</p>
<p>Once the documentation is open, you can search for text.
If you are using the R console, you can use <code>/</code> to open a search bar and
enter-key to search for it.</p>
<hr />
</div>
<div id="back-to-the-file-reading" class="section level4">
<h4><span class="header-section-number">4.2.1.2</span> Back to the file-reading…</h4>
<p>We can use <code>read.delim</code> to read in one of the files.
You will have to specify the delimiter/separator.
" is usually how you specify a TAB character in character strings, like
you would set as that argument.</p>
<div class="incremental">
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb339-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_100.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb339-2" title="2">    <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F)</a>
<a class="sourceLine" id="cb339-3" title="3">viral_protein_data</a></code></pre></div>
<pre><code>##                           V1
## 1 Q8V433.1 Membrane protein 
##                                                       V2
## 1 Bovine respiratory coronavirus (strain 98TXSF-110-LUN)
##                                                                                                                                                                                                                                       V3
## 1 MSSVTTPAPVYTWTADEAIKFLKEWNFSLGIILLFITVILQFGYTSRSMFVYVIKMIILWLMWPLTIILTIFNCVYALNNVYLGFSIVFTIVAIIMWIVYFVNSIRLFIRTGSWWSFNPETNNLMCIDMKGRMYVRPIIEDYHTLTVTIIRGHLYMQGIKLGTGYSLSDLPAYVTVAKVSHLLTYKRGFLDKIGDTSGFAVYVKSKVGNYRLPSTQKGSGLDTALLRNNI</code></pre>
<p>What type is that third column?</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb341-1" title="1"><span class="kw">is</span>(viral_protein_data<span class="op">$</span>V3)</a></code></pre></div>
<pre><code>## [1] &quot;factor&quot;              &quot;integer&quot;             &quot;oldClass&quot;           
## [4] &quot;double&quot;              &quot;numeric&quot;             &quot;vector&quot;             
## [7] &quot;data.frameRowLabels&quot;</code></pre>
<p>Huh, looks like it’ll convert it to a factor automatically. Let’s tell it to
not do that, to leave it <code>as.is</code>.</p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb343-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_100.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb343-2" title="2">    <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F,<span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb343-3" title="3">viral_protein_data</a></code></pre></div>
<pre><code>##                           V1
## 1 Q8V433.1 Membrane protein 
##                                                       V2
## 1 Bovine respiratory coronavirus (strain 98TXSF-110-LUN)
##                                                                                                                                                                                                                                       V3
## 1 MSSVTTPAPVYTWTADEAIKFLKEWNFSLGIILLFITVILQFGYTSRSMFVYVIKMIILWLMWPLTIILTIFNCVYALNNVYLGFSIVFTIVAIIMWIVYFVNSIRLFIRTGSWWSFNPETNNLMCIDMKGRMYVRPIIEDYHTLTVTIIRGHLYMQGIKLGTGYSLSDLPAYVTVAKVSHLLTYKRGFLDKIGDTSGFAVYVKSKVGNYRLPSTQKGSGLDTALLRNNI</code></pre>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb345-1" title="1"><span class="kw">is</span>(viral_protein_data<span class="op">$</span>V3)</a></code></pre></div>
<pre><code>## [1] &quot;character&quot;           &quot;vector&quot;              &quot;data.frameRowLabels&quot;
## [4] &quot;SuperClassMethod&quot;</code></pre>
<p>This is how you do it in base R. How do you do this in tidyverse?
You can use <code>read_delim</code> or <code>read_tsv</code>, with different arguments,
such as <code>col_names=F</code> to prevent it from reading the first line as a header.
Also, no factor conversion :)</p>
</div>
<hr />
</div>
</div>
<div id="string-stuff" class="section level3">
<h3><span class="header-section-number">4.2.2</span> String stuff</h3>
<p>Once we have that, we can test how we will process the protein sequence.
For now, let’s just calculate the length of the protein sequence.</p>
<p>Character strings are a common type of data, and there’s lots of ways to
cut, dice, extract, and recognize elements to help your work.
Here’s some tools/ideas:</p>
<div id="base-r-tools" class="section level4">
<h4><span class="header-section-number">4.2.2.1</span> Base R tools</h4>
<p>Let’s start with one of those above sentences.
You can make a string by putting characters in between single or double quotes:</p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb347-1" title="1">stringz &lt;-<span class="st"> &quot;Character strings are a common type of data, and there&#39;s lots of ways to&quot;</span></a>
<a class="sourceLine" id="cb347-2" title="2">stringz</a></code></pre></div>
<pre><code>## [1] &quot;Character strings are a common type of data, and there&#39;s lots of ways to&quot;</code></pre>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb349-1" title="1"><span class="kw">is</span>(stringz)</a></code></pre></div>
<pre><code>## [1] &quot;character&quot;           &quot;vector&quot;              &quot;data.frameRowLabels&quot;
## [4] &quot;SuperClassMethod&quot;</code></pre>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb351-1" title="1"><span class="kw">str</span>(stringz)</a></code></pre></div>
<pre><code>##  chr &quot;Character strings are a common type of data, and there&#39;s lots of ways to&quot;</code></pre>
<p><code>grep</code> and <code>grepl</code> are good for searching for patterns (like <code>grep</code> in bash).
The first returns position, the second returns <code>TRUE</code> or <code>FALSE</code>.</p>
<p>I only ever use <code>grepl</code> anymore. It searches for a <code>pattern</code> string
in a character vector <code>x</code>, and returns a logical vector.
With a logical vector, you can do a lot, like indexing/slicing or sum-ing.</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb353-1" title="1"><span class="kw">grepl</span>(<span class="dt">pattern=</span><span class="st">&quot;common type&quot;</span>,<span class="dt">x=</span>stringz)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb355-1" title="1"><span class="kw">grepl</span>(<span class="st">&quot;rare type&quot;</span>,stringz)</a></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb357-1" title="1"><span class="kw">grepl</span>(<span class="st">&quot; [comn]* type&quot;</span>,stringz)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>So see that last one with the <code>[]*</code> stuff?
Called a regular expression.
You can use complex
<a href="https://bookdown.org/rdpeng/rprogdatascience/regular-expressions.html">regular expressions</a>
to specify more complex patterns.
These can get
<a href="https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions">really complex</a>
and <a href="https://cs.lmu.edu/~ray/notes/regex/">really powerful</a>.
We’re not going to explain these here, except just for the example below.</p>
<p>Ask on Slack later if you’d like more help with something specific.</p>
<p>Also of use is <code>sub</code> and <code>gsub</code>.
These substitute patterns with replacements, for the
first occurrence (<code>sub</code>) or globally (<code>gsub</code>).
These are really handy for modifying data-tables.</p>
<p>For example, let’s say you’d like to split up filenames by dates:</p>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb359-1" title="1">datar &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">stringsAsFactors=</span>F,</a>
<a class="sourceLine" id="cb359-2" title="2">    <span class="dt">filenames=</span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb359-3" title="3">        <span class="st">&quot;210607_pilot_transformation_works.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb359-4" title="4">        <span class="st">&quot;210609_transformation_seems_to_work.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb359-5" title="5">        <span class="st">&quot;210610_failed_gel_images.jpeg&quot;</span></a>
<a class="sourceLine" id="cb359-6" title="6">    )</a>
<a class="sourceLine" id="cb359-7" title="7">)</a>
<a class="sourceLine" id="cb359-8" title="8">datar<span class="op">$</span>dates &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d)_(.*)</span><span class="ch">\\</span><span class="st">.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb359-9" title="9">        <span class="dt">replacement=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">1&quot;</span>,<span class="dt">x=</span>datar<span class="op">$</span>filenames)</a>
<a class="sourceLine" id="cb359-10" title="10">datar<span class="op">$</span>names &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d)_(.*)</span><span class="ch">\\</span><span class="st">.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb359-11" title="11">        <span class="dt">replacement=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">2&quot;</span>,<span class="dt">x=</span>datar<span class="op">$</span>filenames)</a>
<a class="sourceLine" id="cb359-12" title="12">datar</a></code></pre></div>
<pre><code>##                                  filenames  dates                        names
## 1   210607_pilot_transformation_works.jpeg 210607   pilot_transformation_works
## 2 210609_transformation_seems_to_work.jpeg 210609 transformation_seems_to_work
## 3            210610_failed_gel_images.jpeg 210610            failed_gel_images</code></pre>
<p>The above uses regular expressions, where <code>\\d</code> is matching any digit,
<code>.*</code> matches anything, and <code>()</code> denote “groups” to “capture”.
In the “replacement”, these “groups” are referred to by <code>\\1</code> or <code>\\2</code>.</p>
<p>These “capture groups” are <strong>super useful</strong>!
You can use them to take out matches and re-arrange them.</p>
<p>So, <code>(\\d\\d\\d\\d\\d\\d)_</code> matches six digits followed by an underscore,
such as <code>210618_</code>, and saves it as <code>\\1</code>. So</p>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb361-1" title="1"><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d)_(.*)&quot;</span>,<span class="dt">replacement=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">1&quot;</span>,<span class="dt">x=</span><span class="st">&quot;210616_after&quot;</span>)</a></code></pre></div>
<pre><code>## [1] &quot;210616&quot;</code></pre>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb363-1" title="1"><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d)_(.*)&quot;</span>,<span class="dt">replacement=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">2&quot;</span>,<span class="dt">x=</span><span class="st">&quot;210616_after&quot;</span>)</a></code></pre></div>
<pre><code>## [1] &quot;after&quot;</code></pre>
<p>There are better (cleaner) ways of doing this with tidyverse,
basics are important. See a bit of tidyverse/stringr stuff in the end of this
4.2 section, if you’re curious.</p>
<p>You should now be able to manipulate within character strings a bit,
identify and replace patterns.</p>
<hr />
</div>
</div>
<div id="loops---doing-a-similar-task-multiple-times" class="section level3">
<h3><span class="header-section-number">4.2.3</span> Loops - doing a similar task multiple times</h3>
<p>You will want to repeat this analysis for all the files.
The simplest way of doing this is to copy and paste it, and change the filename.</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb365-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_002.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb365-2" title="2">      <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F, <span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb365-3" title="3"><span class="kw">nchar</span>(viral_protein_data<span class="op">$</span>V3)</a></code></pre></div>
<pre><code>## [1] 177</code></pre>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb367-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_003.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb367-2" title="2">      <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F, <span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb367-3" title="3"><span class="kw">nchar</span>(viral_protein_data[<span class="dv">1</span>,<span class="st">&quot;V3&quot;</span>])</a></code></pre></div>
<pre><code>## [1] 137</code></pre>
<p>Go ahead and do this for all 242 proteins</p>
<hr />
<p>… just kidding.</p>
<p>This is a lot of work, and each time we do this we can introduce errors.
If we ever have more files, we have to copy and paste more. If we ever want
to change an argument for all of them, we have to do each one.</p>
<p>Instead, we can work with a list of all the files available:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb369-1" title="1"><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot; &quot;viral_proteins_001.tsv&quot; &quot;viral_proteins_002.tsv&quot;
## [4] &quot;viral_proteins_003.tsv&quot; &quot;viral_proteins_004.tsv&quot;</code></pre>
<p>Note that I put a <code>[1:5]</code> to limit it to the first 5.
You could also use <code>head()</code>. It’s a good idea to work
with a small subset of files while you are iterating
through development, then scale it up to the entirety.</p>
<p>But watch out…</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb371-1" title="1"><span class="kw">tail</span>(<span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_238.tsv&quot;          &quot;viral_proteins_239.tsv&quot;         
## [3] &quot;viral_proteins_240.tsv&quot;          &quot;viral_proteins_241.tsv&quot;         
## [5] &quot;viral_proteins_242.tsv&quot;          &quot;viral_structural_proteins.fasta&quot;</code></pre>
<p>At the very end is another different file, a FASTA file.
So let’s use a <code>pattern</code>
argument in <code>list.files</code> to specify what we want to list.
And we’ll get back to the FASTA later…</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb373-1" title="1"><span class="kw">tail</span>(<span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,<span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_237.tsv&quot; &quot;viral_proteins_238.tsv&quot; &quot;viral_proteins_239.tsv&quot;
## [4] &quot;viral_proteins_240.tsv&quot; &quot;viral_proteins_241.tsv&quot; &quot;viral_proteins_242.tsv&quot;</code></pre>
<p><code>.*</code> stands for 0 or more (<code>*</code>) of anything (<code>.</code>). For more details,
look up regular-expressions.</p>
<hr />
<div id="what-are-loops" class="section level4">
<h4><span class="header-section-number">4.2.3.1</span> What are loops?</h4>
<!-- video -->
<p>Loops are for running a “code block” as many times
as the “condition” determines.</p>
<ul>
<li><p>A “code block” is either one line of code, or multiple
lines of code surrounded by curly brackets - {}</p>
<pre><code>  {
      code &lt;- &quot;in a block&quot; 
      with &lt;- &quot;multiple lines&quot; 
  }</code></pre>
<p>The code just runs. Yep. It’s that simple.</p></li>
<li><p>A “condition” is an expression of code that
can either evaluate to either <code>TRUE</code> or <code>FALSE</code>, or
set a variable for each time running the code block.
This is often just before the code block, in <code>()</code> parentheses.</p></li>
</ul>
<p>The most common form of these kind of “control statements” is a <code>for</code> loop.
Other “control statements” or “flow control statements” are
<code>while</code>, <code>repeat</code>, and <code>if</code>.</p>
<p>Let’s look up what they do, with <code>?`for`</code>
Note the back ticks! These are a trick in R to make anything
be interpreted as literally what you type, and not
any special characters. Like <code>?`+`</code>, or <code>?`?`</code></p>
<hr />
<p>Here’s an example loop:</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb376-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>) {</a>
<a class="sourceLine" id="cb376-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb376-3" title="3">}</a></code></pre></div>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4</code></pre>
<p>The pieces:</p>
<ul>
<li><p><code>(i in 1:4)</code> is what is being looped over - <code>1:4</code> is a vector of
1 through 4 that is created, and it is put one at a time into <code>i</code>
(a new variable). You need the parentheses.</p></li>
<li><p><code>{</code> and <code>}</code> denote the opening and closing brackets, specify the
“code block” that is run each time.</p></li>
<li><p>inside this “code block” is <code>print(i)</code> - it prints the variable <code>i</code>,
which is set to a value of 1, 2, 3, or 4 for each loop</p></li>
</ul>
<p><a href="http://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth/index.html">More about loops</a></p>
<div class="incremental">
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb378-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb378-2" title="2">        <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb378-3" title="3">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb378-4" title="4">}</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot;
## [1] &quot;viral_proteins_001.tsv&quot;
## [1] &quot;viral_proteins_002.tsv&quot;
## [1] &quot;viral_proteins_003.tsv&quot;
## [1] &quot;viral_proteins_004.tsv&quot;</code></pre>
<p>Inside the code block we can do more than print, like looking for a motif.
Here, we’ll <em>try</em> to read the file and take a look at the sequence …</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb380-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb380-2" title="2">        <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb380-3" title="3">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb380-4" title="4">  <span class="kw">print</span>(<span class="kw">read.delim</span>(i)<span class="op">$</span>V3)</a>
<a class="sourceLine" id="cb380-5" title="5">}</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot;</code></pre>
<pre><code>## Warning in file(file, &quot;rt&quot;): cannot open file &#39;viral_proteins_000.tsv&#39;: No such
## file or directory</code></pre>
<pre><code>## Error in file(file, &quot;rt&quot;): cannot open the connection</code></pre>
</div>
<p>Error! It is looking for a file <code>viral_proteins_000.tsv</code>, but it is looking
in <em>this directory</em>. It is actually in <code>data/viral_structural_proteins</code>.
Look up the <code>list.files</code> documentation, and find how to get it to return
the full name of the file.</p>
<hr />
<p>Next, this should work…</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb384-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb384-2" title="2">        <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>,<span class="dt">full.names=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb384-3" title="3">  <span class="kw">print</span>(<span class="kw">read.delim</span>(i,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F,<span class="dt">as.is=</span>T)<span class="op">$</span>V3)</a>
<a class="sourceLine" id="cb384-4" title="4">}</a></code></pre></div>
<pre><code>## [1] &quot;MEFIPTQTFYNRRYQPRPWTPRPTIQVIRPRPRPQRQAGQLAQLISAVNKLTMRAVPQQKPRRNRKNKKQKQKQQAPQNNTNQKKQPPKKKPAQKKKKPGRRERMCMKIENDCIFEVKHEGKVTGYACLVGDKVMKPAHVKGTIDNADLAKLAFKRSSKYDLECAQIPVHMKSDASKFTHEKPEGYYNWHHGAVQYSGGRFTIPTGAGKPGDSGRPIFDNKGRVVAIVLGGANEGARTALSVVTWNKDIVTKITPEGAEEWSLAIPVMCLLANTTFPCSQPPCIPCCYEKEPEETLRMLEDNVMRPGYYQLLQASLTCSPHRQRRSTKDNFNVYKATRPYLAHCPDCGEGHSCHSPVALERIRNEATDGTLKIQVSLQIGIGTDDSHDWTKLRYMDNHIPADAGRAGLFVRTSAPCTITGTMGHFILARCPKGETLTVGFTDSRKISHSCTHPFHHDPPVIGREKFHSRPQHGKELPCSTYVQSNAATAEEIEVHMPPDTPDRTLLSQQSGNVKITVNGRTVRYKCNCGGSNEGLITTDKVINNCKVDQCHAAVTNHKKWQYNSPLVPRNAELGDRKGKIHIPFPLANVTCMVPKARNPTVTYGKNQVIMLLYPDHPTLLSYRSMGEEPNYQEEWVTHKKEVVLTVPTEGLEVTWGNNEPYKYWPQLSANGTAHGHPHEIILYYYELYPTMTVVVVSVASFILLSMVGMAVGMCMCARRRCITPYELTPGATVPFLLSLICCIRTAKAATYQEAAVYLWNEQQPLFWLQALIPLAALIVLCNCLRLLPCCCKTLAFLAVMSIGAHTVSAYEHVTVIPNTVGVPYKTLVNRPGYSPMVLEMELLSVTLEPTLSLDYITCEYKTVIPSPYVKCCGTAECKDKNLPDYSCKVFTGVYPFMWGGAYCFCDAENTQLSEAHVEKSESCKTEFASAYRAHTASASAKLRVLYQGNNITVTAYANGDHAVTVKDAKFIVGPMSSAWTPFDNKIVVYKGDVYNMDYPPFGAGRPGQFGDIQSRTPESKDVYANTQLVLQRPAAGTVHVPYSQAPSGFKYWLKERGASLQHTAPFGCQIATNPVRAMNCAVGNMPISIDIPDAAFTRVVDAPSLTDMSCEVPACTHSSDFGGVAIIKYAVSKKGKCAVHSMTNAVTIREAEIEVEGNSQLQISFSTALASAEFRVQVCSTQVHCAAECHPPKDHIVNYPASHTTLGVQDISATAMSWVQKITGGVGLVVAVAALILIVVLCVSFSRH&quot;
## [1] &quot;MFPFQPMYPMQPMPYRNPFAAPRRPWFPRTDPFLAMQVQELTRSMANLTFKQRRGAPPEGPPAKKSKREAPQKQRGGQRKKKKNEGKKKAKTGPPNLKTQNGNKKKTNKKPGKRQRMVMKLESDKTFPIMLEGKINGYACVVGGKLFRPMHVEGKIDNDVLAALKTKKASKYDLEYADVPQNMRADTFKYTHEKPQGYYSWHHGAVQYENGRFTVPRGVGARGDSGRPILDNQGRVVAIVLGGVNEGSRTALSVVMWNEKGVTVKYTPENCEQWSLVTTMCLLANVTFPCAQPPICYDRKPAETLAMLSANVDNPGYDELLKAAVTCPGRKRRSTEELFKEYKLTRPYMARCVRCAVGSCHSPIAIEAVKSDGHDGYVRLQTSSQYGLDPSGNLKSRTMRYNMYGTIEEIPLHQVSLHTSRPCHIVDGHGYFLLARCPAGDSITMEFKKDSVTHSCSVPYEVKFNPVGRELYTHPPEHGAEQACQVYAHDAQNRGAYVEMHLPGSEVDSSLVSLSSGLVSVTPPAGTSALVECECSGTTISKTINKTKQFSQCTKKEQCRAYRLQNDKWVYNSDKLPKAAGATLKGKLHVPFLLADGKCTVPLAPEPMITFGFRSVSLKLHPKYPTYLTTRELADEPHYTHELISEPSVRNFSVTAKGWEFVWGNHPPKRFWAQETAPGNPHGLPHEVIVHYYHRYPMSTITGLSICAAIVAVSIAASTWLLCRSRASCLTPYRLTPNAKMPLCLAVLCCARSARAETTWESLDHLWNNNQQMFWTQLLIPLAALIVVTRLLKCMCCVVPFLVVAGAAGAGAYEHATTMPNQAGISYNTIVNRAGYAPLPISITPTKIKLIPTVNLEYVTCHYKTGMDSPTIKCCGSQECTPTYRPDEQCKVFAGVYPFMWGGAYCFCDTENTQISKAYVMKSEDCLADHAAAYKAHTASVQALLNITVGEHSTVTTVYVNGETPVNFNGVKLTAGPLSTAWTPFDRKIVQYAGEIYNYDFPEYGAGQPGAFGDIQLRTVSSSDLYANTNLVLQRPKAGAIHVPYTQAPSGFEQWKKDKAPSLKFTAPFGCEIYTNPIRAENCAVGSIPLAFDIPDALFTRVSETPTLSAAECTLNECVYSSDFGGIATVKYSASKSGKCAVHVPSGTATLKEASVELAEQGSVTIHFSTANIHPEFRLQICTSFVTCKGDCHPPKDHIVTHPQYHAQTFTAAVSKTAWTWLTSLLGGSAVIIIIGLVLATLVAMYVLTNQKHN&quot;
## [1] &quot;MFNIKMTISTLLIALIILVIIILVVFLYYKKQQPPKKVCKVDKDCGSGEHCVRGTCSTLSCLDAVKMDKRNIKIDSKISSCEFTPNFYRFTDTAADEQQEFGKTRHPIKITPSPSESHSPQEVCEKYCSWGTDDCTGWEYVGDEKEGTCYVYNNPHHPVLKYGKDHIIALPRNHKHA&quot;
## [1] &quot;MEAVLTKLDQEEKKALQNFHRCAWEETKNIINDFLEIPEERCTYKFNSYTKKMELLFTPEFHTAWHEVPECREFILNFLRLISGHRVVLKGPTFVFTKETKNLGIPSTINVDFQANIENMDDLQKGNLIGKMNIKEG&quot;
## [1] &quot;MAFLMSEFIGLGLAGAGVLSNALLRRQELQLQKQALENGLVLKADQLGRLGFNPNEVKNVIVGNSFSSNVRLSNMHNDASVVNAYNVYNPASNGIRKKIKSLNNSVKIYNTTGESSV&quot;</code></pre>
<p>And yep, we have protein sequences.
Now looking for some amino acids:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb386-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb386-2" title="2">        <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>,<span class="dt">full.names=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb386-3" title="3">  <span class="kw">print</span>(<span class="kw">grepl</span>(<span class="st">&quot;QQ&quot;</span>,<span class="kw">read.delim</span>(i,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F,<span class="dt">as.is=</span>T)<span class="op">$</span>V3))</a>
<a class="sourceLine" id="cb386-4" title="4">}</a></code></pre></div>
<pre><code>## [1] TRUE
## [1] TRUE
## [1] TRUE
## [1] FALSE
## [1] FALSE</code></pre>
<hr />
</div>
</div>
<div id="storing-values-from-a-loop" class="section level3">
<h3><span class="header-section-number">4.2.4</span> Storing values from a loop</h3>
<p>First, how can we store the filenames and use them later for the loop?
We need to turn out list of files into indicies.
We’ll save it first so we can count how many there are.</p>
<p><code>seq_along</code> is handy function to create a number sequence along a vector,
otherwise use something like <code>seq(1,length(x))</code>.</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb388-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb388-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,<span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb388-3" title="3"></a>
<a class="sourceLine" id="cb388-4" title="4"><span class="kw">length</span>(first_five)</a></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb390-1" title="1"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb390-2" title="2"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb390-3" title="3">    <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb390-4" title="4">    <span class="kw">print</span>(first_five[i])</a>
<a class="sourceLine" id="cb390-5" title="5">}</a></code></pre></div>
<pre><code>## [1] 1
## [1] &quot;viral_proteins_000.tsv&quot;
## [1] 2
## [1] &quot;viral_proteins_001.tsv&quot;
## [1] 3
## [1] &quot;viral_proteins_002.tsv&quot;
## [1] 4
## [1] &quot;viral_proteins_003.tsv&quot;
## [1] 5
## [1] &quot;viral_proteins_004.tsv&quot;</code></pre>
<hr />
<p>Okay, but how do we store values for later analysis?</p>
<p>In other languages, “append”. But R is not built that way. It’ll work
(using something like <code>append</code>, or <code>x &lt;- c(x, new_value)</code>),
but it’s inefficient. The “R-way” to store values from a loop
is to define a vector of the right length, then put each element in it.</p>
<p>Here’s some example vectors that we can create.</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb392-1" title="1"><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="dv">10</span>)</a></code></pre></div>
<pre><code>##  [1] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot;</code></pre>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb394-1" title="1"><span class="kw">vector</span>(<span class="dt">mode=</span><span class="st">&quot;numeric&quot;</span>,<span class="dt">length=</span><span class="dv">5</span>)</a></code></pre></div>
<pre><code>## [1] 0 0 0 0 0</code></pre>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb396-1" title="1"><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="dv">2</span>)</a></code></pre></div>
<pre><code>## [1] FALSE FALSE</code></pre>
<p>You can save these to a variable, and thus have a character vector,
numeric vector, or logical vector, of different sizes.</p>
<hr />
<p>Putting these together, we can create and save a vector of file names:</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb398-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb398-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,<span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb398-3" title="3"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb398-4" title="4">filenamez &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb398-5" title="5"></a>
<a class="sourceLine" id="cb398-6" title="6"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb398-7" title="7"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb398-8" title="8">  filenamez[i] &lt;-<span class="st"> </span>first_five[i]</a>
<a class="sourceLine" id="cb398-9" title="9">}</a>
<a class="sourceLine" id="cb398-10" title="10"></a>
<a class="sourceLine" id="cb398-11" title="11">filenamez</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot; &quot;viral_proteins_001.tsv&quot; &quot;viral_proteins_002.tsv&quot;
## [4] &quot;viral_proteins_003.tsv&quot; &quot;viral_proteins_004.tsv&quot;</code></pre>
<hr />
<p>And finally calculate the length of each protein:
And</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb400-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb400-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb400-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb400-4" title="4"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb400-5" title="5">has_qq &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb400-6" title="6"></a>
<a class="sourceLine" id="cb400-7" title="7"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb400-8" title="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb400-9" title="9">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb400-10" title="10">  has_qq[i] &lt;-<span class="st"> </span><span class="kw">grepl</span>( <span class="st">&quot;QQ&quot;</span>, <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb400-11" title="11">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3</a>
<a class="sourceLine" id="cb400-12" title="12">    )</a>
<a class="sourceLine" id="cb400-13" title="13">}</a>
<a class="sourceLine" id="cb400-14" title="14"></a>
<a class="sourceLine" id="cb400-15" title="15">has_qq</a></code></pre></div>
<pre><code>## [1]  TRUE  TRUE  TRUE FALSE FALSE</code></pre>
<hr />
<p>Now we can take off the <code>[1:5]</code> limiter, and do the whole set:</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb402-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb402-2" title="2">filez &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb402-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb402-4" title="4"></a>
<a class="sourceLine" id="cb402-5" title="5"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb402-6" title="6">has_qq &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="kw">length</span>(filez))</a>
<a class="sourceLine" id="cb402-7" title="7"></a>
<a class="sourceLine" id="cb402-8" title="8"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb402-9" title="9"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(filez) ) {</a>
<a class="sourceLine" id="cb402-10" title="10">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb402-11" title="11">  has_qq[i] &lt;-<span class="st"> </span><span class="kw">grepl</span>( <span class="st">&quot;QQ&quot;</span>, <span class="kw">read.delim</span>(filez[i],</a>
<a class="sourceLine" id="cb402-12" title="12">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3</a>
<a class="sourceLine" id="cb402-13" title="13">    )</a>
<a class="sourceLine" id="cb402-14" title="14">}</a>
<a class="sourceLine" id="cb402-15" title="15"></a>
<a class="sourceLine" id="cb402-16" title="16">has_qq[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>##  [1]  TRUE  TRUE  TRUE FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE</code></pre>
<hr />
<p>How many proteins have “QQ”? Cool trick for logical vectors, <code>sum</code>.</p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb404-1" title="1"><span class="kw">sum</span>(has_qq)</a></code></pre></div>
<pre><code>## [1] 148</code></pre>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb406-1" title="1"><span class="kw">table</span>(has_qq)</a></code></pre></div>
<pre><code>## has_qq
## FALSE  TRUE 
##    95   148</code></pre>
<hr />
<p>Can also ggplot it, for fun:</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb408-1" title="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb408-2" title="2"><span class="kw">ggplot</span>( <span class="kw">data.frame</span>(<span class="dt">qq_status=</span>has_qq) )<span class="op">+</span></a>
<a class="sourceLine" id="cb408-3" title="3"><span class="st">    </span><span class="kw">theme_classic</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb408-4" title="4"><span class="st">    </span><span class="kw">aes</span>(<span class="dt">x=</span>qq_status)<span class="op">+</span></a>
<a class="sourceLine" id="cb408-5" title="5"><span class="st">    </span><span class="kw">geom_bar</span>()</a></code></pre></div>
<p><img src="520-viral-protein-analysis_files/figure-html/ggplot_viz-1.png" width="672" /></p>
<hr />
</div>
<div id="extending-the-workflow" class="section level3">
<h3><span class="header-section-number">4.2.5</span> Extending the workflow</h3>
<p>How do you think about/plan the workflow we’ve built?</p>
<p>How is it organized?</p>
<p>One way is to flatten out all the tasks,
and to script each individual task every time it is done.
This requires the author, user, and reader to understand a lot of complexity.</p>
<div class="incremental">
<div class="mermaid">
graph TD
    classDef default line-height:12px;
    classDef three line-height:12px,fill:yellow;
    you[analyst] --> B[finding files] &    C[making a vector] &    D[looping] &    E[reading files] &    F[getting protein seq] &    G[calculting and saving nchar] &    H[plotting nchar]
    B:::three;
    C:::three;
    D:::three;
    E:::three;
    F:::three;
    G:::three;
    H:::three;
</div>
<p>Another way to approach this is to cluster them into a hierarchy of modules.</p>
<div class="mermaid">
graph TD
    classDef default line-height:12px;
    classDef one  line-height:12px,fill:pink;
    classDef two line-height:12px,fill:cyan;
    classDef three line-height:12px,fill:yellow;
    you2[analyst] --> find & read & plot;
    find:::one --> B;
    B[finding files]:::three;
    read:::one --> looping & reading;
    looping:::two --> C & D;
    C[making a vector]:::three;
    D[looping]:::three;
    reading[reading and calc]:::two --> E & F & G;
    E[reading files]:::three;
    F[getting protein seq]:::three;
    G[calculting and saving nchar]:::three;
    plot:::one --> H;
    H[plotting nchar]:::three;
</div>
<p>In this organization, the analyst can operate at levels of
<span style="color:magenta">steps</span>,
<span style="color:cyan">modules</span>, and
<span style="color:gold">specific instructions</span>,
depending on what is needed.</p>
</div>
<hr />
<p>Organizing your workflows into composable modules lets you extend these to
un-ancipiated complexity. You could imaging using these steps or modules:</p>
<div class="mermaid">
graph LR
    classDef default line-height:12px;
    vp[read viral proteins] --> cl[calculate lengths] --> hist[histogram]
</div>
<p>in new ways by composing the elements together, to analyze a different source
of proteins, with a new analysis, with similar plots:</p>
<div class="mermaid">
graph LR
    classDef default line-height:12px;
    vp[read viral proteins] --> cl[calculate lengths] --> hist[histogram]
    hp[read human proteins] --> cl --> boxplot[boxplot]
    vp --> cmotif[find a particular motif] --> motifhist[histogram]
    hp --> cmotif --> motifboxplot[boxplot]
</div>
<p>Where maybe these are how the inputs/outputs are defined:</p>
<pre><code>reading_proteins:
    input: directory path of proteins to read
    output: protein sequences
calculating lengths:
    input: protein sequences
    output: numeric vector, of lengths
finding a motif:
    input: protein sequences
    output: numeric vector, of presence/absence</code></pre>
<hr />
</div>
<div id="conclusion" class="section level3">
<h3><span class="header-section-number">4.2.6</span> Conclusion</h3>
<p>Write this up!
Edit your rmarkdown file to use</p>
<div id="headers" class="section level4">
<h4><span class="header-section-number">4.2.6.1</span> headers</h4>
<p><em>italics</em></p>
<p><em>bold statements</em></p>
<p>, reference previous figures, or make additional plots, or list
out filenames to support your claim.
Maybe just spend five minutes for now, but get used to writing your
notes down so you save them.</p>
<p>Ideas:
- You can analyze strings using functions like <code>nchar</code> or <code>grepl</code>.
- You can use loops to organize repetitive tasks, such as working through a
large list of files and reading them.
- Modular design of workflows, using functions or just organized and
commented code, can help you extend your analysis to handle new
questions as they arise</p>
<hr />
</div>
</div>
<div id="totally-optional-supplements-for-your-edification" class="section level3">
<h3><span class="header-section-number">4.2.7</span> Totally optional supplements, for your edification</h3>
<div id="stringr-and-the-tidyverse" class="section level4">
<h4><span class="header-section-number">4.2.7.1</span> stringr and the tidyverse</h4>
<p>The <code>tidyverse</code> includes a package <code>stringr</code>. This cleans up some of these
basic string operations, makes them more standardized, and gives them nice
standardized names. It’s worth using, but you should know about the other
base R functions!</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb410-1" title="1"><span class="kw">library</span>(stringr)</a></code></pre></div>
<p>For example the above,</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb411-1" title="1">datar &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">stringsAsFactors=</span>F,</a>
<a class="sourceLine" id="cb411-2" title="2">    <span class="dt">filenames=</span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb411-3" title="3">        <span class="st">&quot;210607_pilot_transformation_works.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb411-4" title="4">        <span class="st">&quot;210609_transformation_seems_to_work.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb411-5" title="5">        <span class="st">&quot;210610_failed_gel_images.jpeg&quot;</span></a>
<a class="sourceLine" id="cb411-6" title="6">    )</a>
<a class="sourceLine" id="cb411-7" title="7">)</a>
<a class="sourceLine" id="cb411-8" title="8">datar<span class="op">$</span>dates &lt;-<span class="st"> </span><span class="kw">str_extract</span>(<span class="dt">string=</span>datar<span class="op">$</span>filenames,<span class="dt">pattern=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d_&quot;</span>)</a>
<a class="sourceLine" id="cb411-9" title="9">datar</a></code></pre></div>
<pre><code>##                                  filenames   dates
## 1   210607_pilot_transformation_works.jpeg 210607_
## 2 210609_transformation_seems_to_work.jpeg 210609_
## 3            210610_failed_gel_images.jpeg 210610_</code></pre>
<p>There are many options, type <code>str_</code> and then TAB.</p>
<p>For much much more about strings, check out
<a href="https://r4ds.had.co.nz/strings.html">Wickham’s R4DS book</a>.
Especially for the special characters part</p>
<p>Or even cleaner using <code>tidyverse</code> verbs:</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb413-1" title="1"><span class="kw">library</span>(tidyverse)</a></code></pre></div>
<pre><code>## ── Attaching packages ───────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>## ✔ tibble  3.0.4     ✔ purrr   0.3.4
## ✔ tidyr   1.1.2     ✔ dplyr   1.0.2
## ✔ readr   1.4.0     ✔ forcats 0.5.0</code></pre>
<pre><code>## ── Conflicts ──────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb417-1" title="1"><span class="kw">tibble</span>(<span class="dt">filenames=</span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb417-2" title="2">            <span class="st">&quot;210607_pilot_transformation_works.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb417-3" title="3">            <span class="st">&quot;210609_transformation_seems_to_work.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb417-4" title="4">            <span class="st">&quot;210610_failed_gel_images.jpeg&quot;</span></a>
<a class="sourceLine" id="cb417-5" title="5">        )</a>
<a class="sourceLine" id="cb417-6" title="6">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb417-7" title="7"><span class="st">    </span><span class="kw">separate</span>(<span class="dt">col=</span>filenames,<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,<span class="dt">into=</span><span class="kw">c</span>(<span class="st">&quot;date&quot;</span>,<span class="st">&quot;rest&quot;</span>),<span class="dt">extra=</span><span class="st">&quot;merge&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb417-8" title="8"><span class="st">    </span><span class="kw">separate</span>(<span class="dt">col=</span>rest,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">.&quot;</span>,<span class="dt">into=</span><span class="kw">c</span>(<span class="st">&quot;name&quot;</span>,<span class="st">&quot;type&quot;</span>)) </a></code></pre></div>
<pre><code>## # A tibble: 3 x 3
##   date   name                         type 
##   &lt;chr&gt;  &lt;chr&gt;                        &lt;chr&gt;
## 1 210607 pilot_transformation_works   jpeg 
## 2 210609 transformation_seems_to_work jpeg 
## 3 210610 failed_gel_images            jpeg</code></pre>
<p>I’d recommend you use the <code>tidyverse</code>-style manipulations - they’re powerful.
But you should be familiar with base R, to fix bugs.</p>
<hr />
</div>
<div id="more-tidyverse-demo" class="section level4">
<h4><span class="header-section-number">4.2.7.2</span> More <code>tidyverse</code> demo</h4>
<p><code>tidyverse</code> is pretty neat at doing these things in simpler, more readable
code. Here’s how you might approach the motif-finding task in that paradigm:</p>
<p>1st - list files, but set it in a tibble:</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb419-1" title="1"><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb419-2" title="2">    <span class="dt">filenames=</span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb419-3" title="3">        <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>) </a>
<a class="sourceLine" id="cb419-4" title="4">    )</a></code></pre></div>
<pre><code>## # A tibble: 243 x 1
##    filenames                                            
##    &lt;chr&gt;                                                
##  1 data/viral_structural_proteins/viral_proteins_000.tsv
##  2 data/viral_structural_proteins/viral_proteins_001.tsv
##  3 data/viral_structural_proteins/viral_proteins_002.tsv
##  4 data/viral_structural_proteins/viral_proteins_003.tsv
##  5 data/viral_structural_proteins/viral_proteins_004.tsv
##  6 data/viral_structural_proteins/viral_proteins_005.tsv
##  7 data/viral_structural_proteins/viral_proteins_006.tsv
##  8 data/viral_structural_proteins/viral_proteins_007.tsv
##  9 data/viral_structural_proteins/viral_proteins_008.tsv
## 10 data/viral_structural_proteins/viral_proteins_009.tsv
## # … with 233 more rows</code></pre>
<p>2nd - pipe that into a <code>mutate</code> call, where you use <code>map</code> on filenames.</p>
<p><code>map</code> is like <code>lapply</code>. It takes whatever you have on the left, and puts
it one by one into the function on the right, along with optional arguments
that you list after it. Here, we set <code>col_names</code> to false,
analogous to <code>header</code> in <code>read.delim</code>.</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb421-1" title="1"><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb421-2" title="2">    <span class="dt">filenames=</span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb421-3" title="3">        <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>) </a>
<a class="sourceLine" id="cb421-4" title="4">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb421-5" title="5"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">rawfiles=</span><span class="kw">map</span>(filenames,read_tsv,<span class="dt">col_names=</span>F)) </a></code></pre></div>
<pre><code>## # A tibble: 243 x 2
##    filenames                                             rawfiles        
##    &lt;chr&gt;                                                 &lt;list&gt;          
##  1 data/viral_structural_proteins/viral_proteins_000.tsv &lt;tibble [1 × 3]&gt;
##  2 data/viral_structural_proteins/viral_proteins_001.tsv &lt;tibble [1 × 3]&gt;
##  3 data/viral_structural_proteins/viral_proteins_002.tsv &lt;tibble [1 × 3]&gt;
##  4 data/viral_structural_proteins/viral_proteins_003.tsv &lt;tibble [1 × 3]&gt;
##  5 data/viral_structural_proteins/viral_proteins_004.tsv &lt;tibble [1 × 3]&gt;
##  6 data/viral_structural_proteins/viral_proteins_005.tsv &lt;tibble [1 × 3]&gt;
##  7 data/viral_structural_proteins/viral_proteins_006.tsv &lt;tibble [1 × 3]&gt;
##  8 data/viral_structural_proteins/viral_proteins_007.tsv &lt;tibble [1 × 3]&gt;
##  9 data/viral_structural_proteins/viral_proteins_008.tsv &lt;tibble [1 × 3]&gt;
## 10 data/viral_structural_proteins/viral_proteins_009.tsv &lt;tibble [1 × 3]&gt;
## # … with 233 more rows</code></pre>
<p>3rd - we then <code>unnest</code> the rawfiles. This takes a column of tibbles and makes
it magically just line up with the top level tibble.</p>
<p>Now we’ll also save it as variable <code>datar</code>.</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb423-1" title="1">datar &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb423-2" title="2">    <span class="dt">filenames=</span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb423-3" title="3">        <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>) </a>
<a class="sourceLine" id="cb423-4" title="4">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb423-5" title="5"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">rawfiles=</span><span class="kw">map</span>(filenames,read_tsv,<span class="dt">col_names=</span>F)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb423-6" title="6"><span class="st">    </span><span class="kw">unnest</span>(rawfiles)</a></code></pre></div>
<p>Then we can process on <code>datar</code>. Note the use of <code>mutate</code>, <code>map</code>, and
wrapping <code>map</code> in unlist to just return the lengths instead of a list of
lengths.</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb424-1" title="1">datar <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb424-2" title="2"><span class="st">    </span><span class="kw">mutate</span>(</a>
<a class="sourceLine" id="cb424-3" title="3">        <span class="dt">protein_length=</span><span class="kw">unlist</span>(<span class="kw">map</span>(X3,nchar)),</a>
<a class="sourceLine" id="cb424-4" title="4">        <span class="dt">qq_present=</span><span class="kw">unlist</span>(<span class="kw">map</span>(X3,grepl,<span class="dt">pattern=</span><span class="st">&quot;QQ&quot;</span>)),</a>
<a class="sourceLine" id="cb424-5" title="5">    )</a></code></pre></div>
<pre><code>## # A tibble: 243 x 6
##    filenames       X1         X2         X3            protein_length qq_present
##    &lt;chr&gt;           &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;                  &lt;int&gt; &lt;lgl&gt;     
##  1 data/viral_str… Q8JUX5.3 … Chikungun… MEFIPTQTFYNR…           1248 TRUE      
##  2 data/viral_str… YP_009507… Everglade… MFPFQPMYPMQP…           1254 TRUE      
##  3 data/viral_str… NP_042704… African s… MFNIKMTISTLL…            177 TRUE      
##  4 data/viral_str… NP_042736… African s… MEAVLTKLDQEE…            137 FALSE     
##  5 data/viral_str… P27413.1 … Rabbit he… MAFLMSEFIGLG…            117 FALSE     
##  6 data/viral_str… NP_042800… African s… MDTETSPLLSHN…            117 TRUE      
##  7 data/viral_str… NP_050192… Human bet… MDLKAQSIPFAW…            858 FALSE     
##  8 data/viral_str… Q306W5.1 … Eastern e… MFPYPTLNYSPM…           1242 FALSE     
##  9 data/viral_str… NP_040824… Venezuela… MFPFQPMYPMQP…           1255 TRUE      
## 10 data/viral_str… YP_003987… Acanthamo… MNYSLEDLPNSG…            234 FALSE     
## # … with 233 more rows</code></pre>
<p>And then we can pipe it into a ggplot:</p>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb426-1" title="1">datar <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb426-2" title="2"><span class="st">    </span><span class="kw">mutate</span>(</a>
<a class="sourceLine" id="cb426-3" title="3">        <span class="dt">protein_length=</span><span class="kw">unlist</span>(<span class="kw">map</span>(X3,nchar)),</a>
<a class="sourceLine" id="cb426-4" title="4">        <span class="dt">qq_present=</span><span class="kw">unlist</span>(<span class="kw">map</span>(X3,grepl,<span class="dt">pattern=</span><span class="st">&quot;QQ&quot;</span>)),</a>
<a class="sourceLine" id="cb426-5" title="5">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb426-6" title="6"><span class="st">    </span><span class="kw">ggplot</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb426-7" title="7"><span class="st">    </span><span class="kw">aes</span>(<span class="dt">x=</span>qq_present)<span class="op">+</span></a>
<a class="sourceLine" id="cb426-8" title="8"><span class="st">    </span><span class="kw">geom_bar</span>()</a></code></pre></div>
<p><img src="520-viral-protein-analysis_files/figure-html/tidy_show_listfilesplot-1.png" width="672" /></p>
<p>And the entire code:</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb427-1" title="1"><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb427-2" title="2">    <span class="dt">filenames=</span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb427-3" title="3">        <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>) </a>
<a class="sourceLine" id="cb427-4" title="4">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb427-5" title="5"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">rawfiles=</span><span class="kw">map</span>(filenames,read_tsv,<span class="dt">col_names=</span>F)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb427-6" title="6"><span class="st">    </span><span class="kw">unnest</span>(rawfiles) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb427-7" title="7"><span class="st">    </span><span class="kw">mutate</span>(</a>
<a class="sourceLine" id="cb427-8" title="8">        <span class="dt">protein_length=</span><span class="kw">unlist</span>(<span class="kw">map</span>(X3,nchar)),</a>
<a class="sourceLine" id="cb427-9" title="9">        <span class="dt">qq_present=</span><span class="kw">unlist</span>(<span class="kw">map</span>(X3,grepl,<span class="dt">pattern=</span><span class="st">&quot;QQ&quot;</span>)),</a>
<a class="sourceLine" id="cb427-10" title="10">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb427-11" title="11"><span class="st">    </span><span class="kw">ggplot</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb427-12" title="12"><span class="st">    </span><span class="kw">aes</span>(<span class="dt">x=</span>qq_present)<span class="op">+</span></a>
<a class="sourceLine" id="cb427-13" title="13"><span class="st">    </span><span class="kw">geom_bar</span>()</a></code></pre></div>
<p><img src="520-viral-protein-analysis_files/figure-html/tidy_show_listfilestogether-1.png" width="672" /></p>
<p><em>I</em> think that’s tidy.</p>
<hr />
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>
mermaid.initialize({
    startOnLoad: false, // don't render initially, instead....
    logLevel: 3, // for debugging
});
function initMermaid(s) { 
  var diagrams = document.querySelectorAll('.mermaid');
  var i;
  for(i=0;i<diagrams.length;i++){
    if(diagrams[i].offsetWidth>0){
      mermaid.init(undefined, diagrams[i]);
    }
  }
}
initMermaid();
</script>

</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="4">
<li id="fn4"><p>One crucial aspect missing is the actual bash commands used. I did not
save them as a script, and I <em>should</em> go back, save them in a script,
and re-run that to make sure I get the same result!<a href="analyzing-viral-protein-sequences.html#fnref4" class="footnote-back">↩</a></p></li>
<li id="fn5"><p>The formats you’re likely going to have to deal with are CSV, TSV,
FASTQ, FASTA, SAM - at least. <a href="analyzing-viral-protein-sequences.html#fnref5" class="footnote-back">↩</a></p></li>
<li id="fn6"><p>
…since we’re just demonstrating R coding and not representing ourselves as
being familiar enough with the field to properly interpret these results.
I have no idea what occurance of this would mean or not mean, or are we
doing any sort of statistics or comparisons to null expectations.<a href="analyzing-viral-protein-sequences.html#fnref6" class="footnote-back">↩</a></p></li>
</ol>
</div>
<p style="margin-top:10em;text-align:center;font-size:small;">
    <a href="https://github.com/darachm/dll-r/blob/main/LICENSE.md">
        Licensed Creative Commons Attribution-NonCommercial-ShareAlike 4.0
    </a>
</p>
<p style="text-align:center;font-size:small;">
    <a href="https://github.com/rstudio/bookdown" target="blank">
        Website typeset with bookdown
    </a>
</p>

<script> 
// This is to look through, everything that's a floating lone double dash,
// delete it. This is left over for stuff that's trying to be pauses for
// the slides
var listz = document.getElementsByTagName("p");
for (var i=0; i < listz.length; i++) { 
    if (listz[i].innerText === "–") { listz[i].innerText = "" } 
} 
</script>
            </section>

          </div>
        </div>
      </div>
<a href="what-is-a-workflow.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="packages-install-and-use.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/darachm/dll-r/edit/main/520-viral-protein-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["DLL 2021 R syllabus site.html"],
"toc": {
"collapse": "subsection"
},
"highlight": "zenburn"
});
});
</script>

</body>

</html>
