<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.1 Analyzing viral protein sequences | DLL 2021, R section</title>
  <meta name="description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="4.1 Analyzing viral protein sequences | DLL 2021, R section" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="github-repo" content="darachm/dll-r" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.1 Analyzing viral protein sequences | DLL 2021, R section" />
  
  <meta name="twitter:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  

<meta name="author" content="mesako" />
<meta name="author" content="Margaret" />
<meta name="author" content="Samson" />
<meta name="author" content="Zac" />
<meta name="author" content="darachm" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"/>
<link rel="next" href="troubleshooting-getting-help-dissecting-problems.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DLL 2021 R workshops</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Workshop Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="workshop-goals.html"><a href="workshop-goals.html"><i class="fa fa-check"></i><b>1.1</b> Workshop goals</a></li>
<li class="chapter" data-level="1.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html"><i class="fa fa-check"></i><b>1.2</b> Structure and resources</a><ul>
<li class="chapter" data-level="1.2.1" data-path="structure-and-resources.html"><a href="structure-and-resources.html#this-syllabus-and-videos"><i class="fa fa-check"></i><b>1.2.1</b> This syllabus and videos</a></li>
<li class="chapter" data-level="1.2.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html#slack-channel"><i class="fa fa-check"></i><b>1.2.2</b> Slack channel</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="workshop-expectations.html"><a href="workshop-expectations.html"><i class="fa fa-check"></i><b>1.3</b> Workshop expectations</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="day-3-intoduction-to-r.html"><a href="day-3-intoduction-to-r.html"><i class="fa fa-check"></i><b>2</b> Day 3 - intoduction to R</a><ul>
<li class="chapter" data-level="2.1" data-path="using-rstudio.html"><a href="using-rstudio.html"><i class="fa fa-check"></i><b>2.1</b> Using RStudio</a><ul>
<li class="chapter" data-level="2.1.1" data-path="using-rstudio.html"><a href="using-rstudio.html#r-scripts-in-rstudio"><i class="fa fa-check"></i><b>2.1.1</b> R Scripts in RStudio</a></li>
<li class="chapter" data-level="2.1.2" data-path="using-rstudio.html"><a href="using-rstudio.html#console-and-terminal"><i class="fa fa-check"></i><b>2.1.2</b> Console and Terminal</a></li>
<li class="chapter" data-level="2.1.3" data-path="using-rstudio.html"><a href="using-rstudio.html#environment-and-history"><i class="fa fa-check"></i><b>2.1.3</b> Environment and History</a></li>
<li class="chapter" data-level="2.1.4" data-path="using-rstudio.html"><a href="using-rstudio.html#files-plots-and-packages"><i class="fa fa-check"></i><b>2.1.4</b> Files, Plots, and Packages</a></li>
<li class="chapter" data-level="2.1.5" data-path="using-rstudio.html"><a href="using-rstudio.html#putting-it-all-together"><i class="fa fa-check"></i><b>2.1.5</b> Putting It All Together</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-rmds.html"><a href="using-rmds.html"><i class="fa fa-check"></i><b>2.2</b> Using Rmds</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Getting started in R</a><ul>
<li class="chapter" data-level="2.3.1" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#getting-help-in-r"><i class="fa fa-check"></i><b>2.3.1</b> Getting help in R</a></li>
<li class="chapter" data-level="2.3.2" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-variables-and-data-types"><i class="fa fa-check"></i><b>2.3.2</b> Using variables and data types</a></li>
<li class="chapter" data-level="2.3.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-functions"><i class="fa fa-check"></i><b>2.3.3</b> Using functions</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html"><i class="fa fa-check"></i><b>2.4</b> Working with vectors</a><ul>
<li class="chapter" data-level="2.4.1" data-path="working-with-vectors.html"><a href="working-with-vectors.html#building-vectors"><i class="fa fa-check"></i><b>2.4.1</b> Building vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="working-with-vectors.html"><a href="working-with-vectors.html#indexing-and-subsetting-vectors"><i class="fa fa-check"></i><b>2.4.2</b> Indexing and subsetting vectors</a></li>
<li class="chapter" data-level="2.4.3" data-path="working-with-vectors.html"><a href="working-with-vectors.html#using-logic-and-logicals"><i class="fa fa-check"></i><b>2.4.3</b> Using logic and logicals</a></li>
<li class="chapter" data-level="2.4.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html#modifying-vectors"><i class="fa fa-check"></i><b>2.4.4</b> Modifying vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html"><i class="fa fa-check"></i><b>2.5</b> Working with dataframes</a><ul>
<li class="chapter" data-level="2.5.1" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#indexing-and-subsetting-dataframes"><i class="fa fa-check"></i><b>2.5.1</b> Indexing and subsetting dataframes</a></li>
<li class="chapter" data-level="2.5.2" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#exploring-dataframes"><i class="fa fa-check"></i><b>2.5.2</b> Exploring dataframes</a></li>
<li class="chapter" data-level="2.5.3" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#building-and-modifying-dataframes"><i class="fa fa-check"></i><b>2.5.3</b> Building and modifying dataframes</a></li>
<li class="chapter" data-level="2.5.4" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#handling-datatypes-in-dataframes"><i class="fa fa-check"></i><b>2.5.4</b> Handling datatypes in dataframes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="day-4-tidyverse-and-visualizations.html"><a href="day-4-tidyverse-and-visualizations.html"><i class="fa fa-check"></i><b>3</b> Day 4 - tidyverse and visualizations</a><ul>
<li class="chapter" data-level="3.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html"><i class="fa fa-check"></i><b>3.1</b> Reading and processing data</a><ul>
<li class="chapter" data-level="3.1.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#loading-a-file-into-r"><i class="fa fa-check"></i><b>3.1.1</b> Loading a file into R</a></li>
<li class="chapter" data-level="3.1.2" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#handling-issues-in-data"><i class="fa fa-check"></i><b>3.1.2</b> Handling issues in data</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="using-tidyverse.html"><a href="using-tidyverse.html"><i class="fa fa-check"></i><b>3.2</b> Using tidyverse</a><ul>
<li class="chapter" data-level="3.2.1" data-path="using-tidyverse.html"><a href="using-tidyverse.html#goals-to-be-reworked-alongside-content"><i class="fa fa-check"></i><b>3.2.1</b> Goals to be reworked alongside content</a></li>
<li class="chapter" data-level="3.2.2" data-path="using-tidyverse.html"><a href="using-tidyverse.html#welcome-to-the-tidyverse"><i class="fa fa-check"></i><b>3.2.2</b> Welcome to the tidyverse!</a></li>
<li class="chapter" data-level="3.2.3" data-path="using-tidyverse.html"><a href="using-tidyverse.html#lets-get-started-with-tidyverse"><i class="fa fa-check"></i><b>3.2.3</b> Let’s get started with Tidyverse</a></li>
<li class="chapter" data-level="3.2.4" data-path="using-tidyverse.html"><a href="using-tidyverse.html#try-tidyverse-on-the-dataset-we-loaded-in-the-previous-section"><i class="fa fa-check"></i><b>3.2.4</b> Try tidyverse on the dataset we loaded in the previous section</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="using-tidyverse-1.html"><a href="using-tidyverse-1.html"><i class="fa fa-check"></i><b>3.3</b> Using tidyverse</a></li>
<li class="chapter" data-level="3.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html"><i class="fa fa-check"></i><b>3.4</b> Making plots with ggplot2</a><ul>
<li class="chapter" data-level="3.4.1" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#getting-started-with-a-ggplot"><i class="fa fa-check"></i><b>3.4.1</b> Getting started with a ggplot</a></li>
<li class="chapter" data-level="3.4.2" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#changing-plot-aesthetics"><i class="fa fa-check"></i><b>3.4.2</b> Changing plot aesthetics</a></li>
<li class="chapter" data-level="3.4.3" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#exploring-simple-plots"><i class="fa fa-check"></i><b>3.4.3</b> Exploring simple plots</a></li>
<li class="chapter" data-level="3.4.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#visualizing-between-groups"><i class="fa fa-check"></i><b>3.4.4</b> Visualizing between groups</a></li>
<li class="chapter" data-level="3.4.5" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#generating-heatmaps"><i class="fa fa-check"></i><b>3.4.5</b> Generating heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html"><i class="fa fa-check"></i><b>3.5</b> Making scientific figures</a><ul>
<li class="chapter" data-level="3.5.1" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#plotting-error-bars"><i class="fa fa-check"></i><b>3.5.1</b> Plotting error bars</a></li>
<li class="chapter" data-level="3.5.2" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#showing-trends-in-data"><i class="fa fa-check"></i><b>3.5.2</b> Showing trends in data</a></li>
<li class="chapter" data-level="3.5.3" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#saving-figures-locally"><i class="fa fa-check"></i><b>3.5.3</b> Saving figures locally</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html"><i class="fa fa-check"></i><b>3.6</b> Applying basic stats</a><ul>
<li class="chapter" data-level="3.6.1" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#describing-quantitative-variables"><i class="fa fa-check"></i><b>3.6.1</b> Describing quantitative variables</a></li>
<li class="chapter" data-level="3.6.2" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#finding-differences-in-data"><i class="fa fa-check"></i><b>3.6.2</b> Finding differences in data</a></li>
<li class="chapter" data-level="3.6.3" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#identifying-correlations"><i class="fa fa-check"></i><b>3.6.3</b> Identifying correlations</a></li>
<li class="chapter" data-level="3.6.4" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#producing-linear-models"><i class="fa fa-check"></i><b>3.6.4</b> Producing linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"><a href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"><i class="fa fa-check"></i><b>4</b> Day 5 - building workflows, using packages, writing (re)usable code, sharing analyses</a><ul>
<li class="chapter" data-level="4.0.1" data-path="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"><a href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html#what-is-a-workflow"><i class="fa fa-check"></i><b>4.0.1</b> What is a workflow?</a></li>
<li class="chapter" data-level="4.1" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html"><i class="fa fa-check"></i><b>4.1</b> Analyzing viral protein sequences</a><ul>
<li class="chapter" data-level="4.1.1" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#reading-in-tab-delimited-file"><i class="fa fa-check"></i><b>4.1.1</b> Reading in tab-delimited file</a></li>
<li class="chapter" data-level="4.1.2" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#loops---doing-a-similar-task-multiple-times"><i class="fa fa-check"></i><b>4.1.2</b> Loops - doing a similar task multiple times</a></li>
<li class="chapter" data-level="4.1.3" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#extending-the-workflow"><i class="fa fa-check"></i><b>4.1.3</b> Extending the workflow</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="troubleshooting-getting-help-dissecting-problems.html"><a href="troubleshooting-getting-help-dissecting-problems.html"><i class="fa fa-check"></i><b>4.2</b> Troubleshooting, getting help, dissecting problems</a><ul>
<li class="chapter" data-level="4.2.1" data-path="troubleshooting-getting-help-dissecting-problems.html"><a href="troubleshooting-getting-help-dissecting-problems.html#places-to-look-for-help"><i class="fa fa-check"></i><b>4.2.1</b> Places to look for help</a></li>
<li class="chapter" data-level="4.2.2" data-path="troubleshooting-getting-help-dissecting-problems.html"><a href="troubleshooting-getting-help-dissecting-problems.html#minimum-reproducible-example"><i class="fa fa-check"></i><b>4.2.2</b> Minimum reproducible example</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="functions-as-modular-steps.html"><a href="functions-as-modular-steps.html"><i class="fa fa-check"></i><b>4.3</b> Functions as modular steps</a><ul>
<li class="chapter" data-level="4.3.1" data-path="functions-as-modular-steps.html"><a href="functions-as-modular-steps.html#tips-for-modular-workflows"><i class="fa fa-check"></i><b>4.3.1</b> Tips for modular workflows</a></li>
<li class="chapter" data-level="4.3.2" data-path="functions-as-modular-steps.html"><a href="functions-as-modular-steps.html#apply-is-a-popular-tool"><i class="fa fa-check"></i><b>4.3.2</b> Apply is a popular tool</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html"><i class="fa fa-check"></i><b>4.4</b> Packages - install and use</a><ul>
<li class="chapter" data-level="4.4.1" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#cran-is-the-canonical-r-package-resource"><i class="fa fa-check"></i><b>4.4.1</b> CRAN is the canonical R package resource</a></li>
<li class="chapter" data-level="4.4.2" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#using-packages-without-library-ing-them"><i class="fa fa-check"></i><b>4.4.2</b> Using packages without <code>library()</code>-ing them</a></li>
<li class="chapter" data-level="4.4.3" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#github-is-a-common-place-for-sharing-packges-in-development"><i class="fa fa-check"></i><b>4.4.3</b> Github is a common place for sharing packges in development</a></li>
<li class="chapter" data-level="4.4.4" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#bioconductor-is-a-great-place-for-bio-related-packages"><i class="fa fa-check"></i><b>4.4.4</b> Bioconductor is a great place for bio-related packages 😬</a></li>
<li class="chapter" data-level="4.4.5" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#what-packages-do-you-have-already"><i class="fa fa-check"></i><b>4.4.5</b> What packages do you have already?</a></li>
<li class="chapter" data-level="4.4.6" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#finding-more-help-and-documentation"><i class="fa fa-check"></i><b>4.4.6</b> Finding more help and documentation</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="writing-reusable-code.html"><a href="writing-reusable-code.html"><i class="fa fa-check"></i><b>4.5</b> Writing (re)usable code</a><ul>
<li class="chapter" data-level="4.5.1" data-path="writing-reusable-code.html"><a href="writing-reusable-code.html#examples"><i class="fa fa-check"></i><b>4.5.1</b> Examples</a></li>
<li class="chapter" data-level="4.5.2" data-path="writing-reusable-code.html"><a href="writing-reusable-code.html#style-guides-can-be-inspiring"><i class="fa fa-check"></i><b>4.5.2</b> Style guides can be inspiring</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html"><i class="fa fa-check"></i><b>4.6</b> More <code>rmarkdown</code> for reproducibile sharing</a><ul>
<li class="chapter" data-level="4.6.1" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html#rmarkdown-package"><i class="fa fa-check"></i><b>4.6.1</b> rmarkdown package</a></li>
<li><a href="more-rmarkdown-for-reproducibile-sharing.html#section"></a></li>
<li class="chapter" data-level="4.6.2" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html#sharing-with-others"><i class="fa fa-check"></i><b>4.6.2</b> Sharing with others</a></li>
<li class="chapter" data-level="4.6.3" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html#reproducibility-and-sharing"><i class="fa fa-check"></i><b>4.6.3</b> Reproducibility and sharing</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>4.7</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="appendix.html"><a href="appendix.html"><i class="fa fa-check"></i><b>5</b> Appendix</a><ul>
<li class="chapter" data-level="5.1" data-path="additional-tutorials.html"><a href="additional-tutorials.html"><i class="fa fa-check"></i><b>5.1</b> Additional tutorials</a><ul>
<li class="chapter" data-level="5.1.1" data-path="additional-tutorials.html"><a href="additional-tutorials.html#sub-sub-section"><i class="fa fa-check"></i><b>5.1.1</b> sub sub section</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="more-challenges.html"><a href="more-challenges.html"><i class="fa fa-check"></i><b>5.2</b> More challenges</a></li>
<li class="chapter" data-level="5.3" data-path="figures-ideas.html"><a href="figures-ideas.html"><i class="fa fa-check"></i><b>5.3</b> Figures ideas</a></li>
<li class="chapter" data-level="5.4" data-path="rmd-example-of-code-chunk.html"><a href="rmd-example-of-code-chunk.html"><i class="fa fa-check"></i><b>5.4</b> Rmd example of code chunk</a></li>
<li class="chapter" data-level="5.5" data-path="more-example-rmds.html"><a href="more-example-rmds.html"><i class="fa fa-check"></i><b>5.5</b> More example Rmds</a></li>
<li class="chapter" data-level="5.6" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html"><i class="fa fa-check"></i><b>5.6</b> How to edit this document</a><ul>
<li class="chapter" data-level="5.6.1" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html#setup"><i class="fa fa-check"></i><b>5.6.1</b> Setup</a></li>
<li class="chapter" data-level="5.6.2" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html#editwork-on-it"><i class="fa fa-check"></i><b>5.6.2</b> Edit/work on it</a></li>
<li class="chapter" data-level="5.6.3" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html#publishing-onto-the-main-page-so-its-hosted-at-the-url"><i class="fa fa-check"></i><b>5.6.3</b> Publishing onto the main page, so it’s hosted at the URL</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="more-links.html"><a href="more-links.html"><i class="fa fa-check"></i><b>5.7</b> More links</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DLL 2021, R section</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="analyzing-viral-protein-sequences" class="section level2">
<h2><span class="header-section-number">4.1</span> Analyzing viral protein sequences</h2>
<p>Let’s work through an analysis of the size of some structural viral proteins.
We’ve done some of the work for you already, and would recommend that you
always document where the data came from and how it was retrieved/processed:</p>
<blockquote>
<p>The NCBI Identital Protein Groups database was queried for “structural”
with “Division” restricted to “Viruses” and “Source database” restricted
to “UniProtKB/Swiss-Prot”. 243 results were retrieved as FASTA files and
converted using bash commands into separate tab-delimited files.
<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a></p>
</blockquote>
<p>These files are in the <code>data/viral_structural_proteins</code> folder,
are tab-delimited three fields, and look like this:</p>
<pre><code>Q8V433.1 Membrane protein       Bovine respiratory coronavirus (strain 98TXSF-110-LUN)  MSSVTTPAPVYTWTA...</code></pre>
<p>How do you inspect these files?
What bash tools do you use?</p>
<hr />
<div id="reading-in-tab-delimited-file" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Reading in tab-delimited file</h3>
<p>Let’s read one of them into R.</p>
<p>The records are “delimited” by tabs, so each field is tab-separated.
We’ll need to use the <code>read.delim</code> function and specify a tab separator.</p>
<div id="review-of-looking-up-documentation" class="section level4">
<h4><span class="header-section-number">4.1.1.1</span> Review of looking up documentation</h4>
<p>You can look up the documentation for any named function or package by using
the <code>?function</code> syntax.</p>
<p>Sometimes, you need to use extra backticks to make it work, like <code>?`+`</code></p>
<p>If you don’t know what you’re looking for, you can search with <code>??</code>.
Extra backticks don’t hurt, and are necessary when you have spaces in the
query.</p>
<p>Once the documentation is open, you can search for text.
If you are using the R console, you can use <code>/</code> to open a search bar and
enter-key to search for it.</p>
<hr />
</div>
<div id="back-to-the-file-reading" class="section level4">
<h4><span class="header-section-number">4.1.1.2</span> Back to the file-reading…</h4>
<p>What does <code>?read.delim</code> say it does?
How do you set the delimiter/separator as a tab?
<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
<p>Is the function expecting there to be a file “header”?
( This would where the first line of the file has the name of each column )
Do we want to set this to <code>TRUE</code> or <code>FALSE</code> ?)</p>
<div class="incremental">
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb309-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_100.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb309-2" title="2">    <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F)</a>
<a class="sourceLine" id="cb309-3" title="3">viral_protein_data</a></code></pre></div>
<pre><code>##                           V1
## 1 Q8V433.1 Membrane protein 
##                                                       V2
## 1 Bovine respiratory coronavirus (strain 98TXSF-110-LUN)
##                                                                                                                                                                                                                                       V3
## 1 MSSVTTPAPVYTWTADEAIKFLKEWNFSLGIILLFITVILQFGYTSRSMFVYVIKMIILWLMWPLTIILTIFNCVYALNNVYLGFSIVFTIVAIIMWIVYFVNSIRLFIRTGSWWSFNPETNNLMCIDMKGRMYVRPIIEDYHTLTVTIIRGHLYMQGIKLGTGYSLSDLPAYVTVAKVSHLLTYKRGFLDKIGDTSGFAVYVKSKVGNYRLPSTQKGSGLDTALLRNNI</code></pre>
<ul>
<li>What is <code>viral_protein_data</code>?</li>
<li>How do we find the column names?</li>
<li>What are two ways to access the protein sequence?</li>
<li>How do we calculate the number of characters in this protein sequence?
( Try searching with “??<code>number of characters</code>” )</li>
</ul>
<p>What type is the <code>V3</code> column? What should it be?
How do we tell R to treat these strings as they are, and not convert to factors?</p>
</div>
<hr />

</div>
</div>
<div id="loops---doing-a-similar-task-multiple-times" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Loops - doing a similar task multiple times</h3>
<p>You will want to repeat this analysis for multiple file.
This automation is great, really useful for scaling your analyses.</p>
<p>Let’s do this for all the proteins.
The simplest way of doing this is to copy and paste it, and change the filename.</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb311-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_002.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb311-2" title="2">      <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F, <span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb311-3" title="3"><span class="kw">nchar</span>(viral_protein_data<span class="op">$</span>V3[[<span class="dv">1</span>]])</a></code></pre></div>
<pre><code>## [1] 177</code></pre>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb313-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_003.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb313-2" title="2">      <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F, <span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb313-3" title="3"><span class="kw">nchar</span>(viral_protein_data[<span class="dv">1</span>,<span class="st">&quot;V3&quot;</span>])</a></code></pre></div>
<pre><code>## [1] 137</code></pre>
<p>Go ahead and do this for all 242 proteins</p>
<hr />
<p>… just kidding.</p>
<p>This is a lot of work, and each time we do this we can introduce errors.
If we ever have more files, we have to copy and paste more. If we ever want
to change an argument for all of them, we have to do each one.</p>
<p>Instead, we can work with a list of all the files available:</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb315-1" title="1"><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot; &quot;viral_proteins_001.tsv&quot; &quot;viral_proteins_002.tsv&quot;
## [4] &quot;viral_proteins_003.tsv&quot; &quot;viral_proteins_004.tsv&quot;</code></pre>
<p>Note that I put a <code>[1:5]</code> to limit it to the first 5.
You could also use <code>head()</code>. It’s a good idea to work
with a small subset of files while you are iterating
through development, then scale it up to the entirety.</p>
<p>Now how do you calculate the number of characters for each protein?</p>
<p>How do we do this for every file listed … ?</p>
<hr />
<div id="what-are-loops" class="section level4">
<h4><span class="header-section-number">4.1.2.1</span> What are loops?</h4>
<!-- video -->
<p>Loops are for running a “code block” as many times
as the “condition” determines.</p>
<ul>
<li><p>A “code block” is either one line of code, or multiple
lines of code surrounded by curly brackets - {}</p>
<pre><code>  {
      code &lt;- &quot;in a block&quot; 
      with &lt;- &quot;multiple lines&quot; 
  }</code></pre>
<p>The code just runs. Yep. It’s that simple.</p></li>
<li><p>A “condition” is an expression of code that
can either evaluate to either <code>TRUE</code> or <code>FALSE</code>, or
set a variable for each time running the code block.
This is often just before the code block, in <code>()</code> parentheses.</p></li>
</ul>
<p>The most common form of these kind of “control statements” is a <code>for</code> loop.
Other “control statements” or “flow control statements” are
<code>while</code>, <code>repeat</code>, and <code>if</code>.</p>
<p>Let’s look up what they do, with <code>?`for`</code>
Note the back ticks! These are a trick in R to make anything
be interpreted as literally what you type, and not
any special characters. Like <code>?`+`</code>, or <code>?`?`</code></p>
<hr />
<p>Here’s an example loop:</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb318-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>) {</a>
<a class="sourceLine" id="cb318-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb318-3" title="3">}</a></code></pre></div>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4</code></pre>
<p>The pieces:</p>
<ul>
<li><p><code>(i in 1:4)</code> is what is being looped over - <code>1:4</code> is a vector of
1 through 4 that is created, and it is put one at a time into <code>i</code>
(a new variable). You need the parentheses.</p></li>
<li><p><code>{</code> and <code>}</code> denote the opening and closing brackets, specify the
“code block” that is run each time.</p></li>
<li><p>inside this “code block” is <code>print(i)</code> - it prints the variable <code>i</code>,
which is set to a value of 1, 2, 3, or 4 for each loop</p></li>
</ul>
<hr />
<p>How do we loop through and print each file name?</p>
<p>What are the (1) code block and (2) loop condition ?</p>
<div class="incremental">
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb320-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb320-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb320-3" title="3">}</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot;
## [1] &quot;viral_proteins_001.tsv&quot;
## [1] &quot;viral_proteins_002.tsv&quot;
## [1] &quot;viral_proteins_003.tsv&quot;
## [1] &quot;viral_proteins_004.tsv&quot;</code></pre>
<p>How do we modify this to calculate the protein length?
I add a line where we use <code>nchar</code>, but <code>i</code> is the filename instead of me
typing it in there.</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb322-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb322-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb322-3" title="3">  <span class="kw">print</span>(<span class="kw">nchar</span>(<span class="kw">read.delim</span>(i)<span class="op">$</span>V3[[<span class="dv">1</span>]]))</a>
<a class="sourceLine" id="cb322-4" title="4">}</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot;</code></pre>
<pre><code>## Warning in file(file, &quot;rt&quot;): cannot open file &#39;viral_proteins_000.tsv&#39;: No such
## file or directory</code></pre>
<pre><code>## Error in file(file, &quot;rt&quot;): cannot open the connection</code></pre>
</div>
<p>Error! It is looking for a file <code>viral_proteins_000.tsv</code>, but it is looking
in <em>this directory</em>. It is actually in <code>data/viral_structural_proteins</code>.
Look up the <code>list.files</code> documentation, and find how to get it to return
the full name of the file.</p>
<hr />
<p>Next, this should work…</p>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb326-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb326-2" title="2">                     <span class="dt">full.names=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb326-3" title="3">  <span class="kw">print</span>(<span class="kw">nchar</span>(<span class="kw">read.delim</span>(i,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F,<span class="dt">as.is=</span>T)<span class="op">$</span>V3[[<span class="dv">1</span>]]))</a>
<a class="sourceLine" id="cb326-4" title="4">}</a></code></pre></div>
<pre><code>## [1] 1248
## [1] 1254
## [1] 177
## [1] 137
## [1] 117</code></pre>
<p>And yep, we have lengths of protein sequences.</p>
<p>…</p>
<p>What now?</p>
<hr />
</div>
<div id="storing-values-from-a-loop" class="section level4">
<h4><span class="header-section-number">4.1.2.2</span> Storing values from a loop</h4>
<p>How do we store these values for later analysis?</p>
<p>In other languages, “append”. But R is not built that way. It’ll work,
but it’s very inefficient. The “R-way” to store values from a loop
is to define a vector of the right length,
then put each element in it.</p>
<p>Here’s some example vectors that we can create.</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb328-1" title="1"><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="dv">10</span>)</a></code></pre></div>
<pre><code>##  [1] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot;</code></pre>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb330-1" title="1"><span class="kw">vector</span>(<span class="dt">mode=</span><span class="st">&quot;numeric&quot;</span>,<span class="dt">length=</span><span class="dv">5</span>)</a></code></pre></div>
<pre><code>## [1] 0 0 0 0 0</code></pre>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb332-1" title="1"><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="dv">2</span>)</a></code></pre></div>
<pre><code>## [1] FALSE FALSE</code></pre>
<hr />
<p>Okay, but how do we access each position to save the value?
We need to turn out list of files into indicies.
We’ll save it first so we can count how many there are.</p>
<p><code>seq_along</code> is handy function to create a number sequence along a vector,
otherwise use something like <code>seq(1,length(x))</code>.</p>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb334-1" title="1">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb334-2" title="2"></a>
<a class="sourceLine" id="cb334-3" title="3"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb334-4" title="4">    <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb334-5" title="5">    <span class="kw">print</span>(first_five[i])</a>
<a class="sourceLine" id="cb334-6" title="6">}</a></code></pre></div>
<pre><code>## [1] 1
## [1] &quot;viral_proteins_000.tsv&quot;
## [1] 2
## [1] &quot;viral_proteins_001.tsv&quot;
## [1] 3
## [1] &quot;viral_proteins_002.tsv&quot;
## [1] 4
## [1] &quot;viral_proteins_003.tsv&quot;
## [1] 5
## [1] &quot;viral_proteins_004.tsv&quot;</code></pre>
<hr />
<p>Putting these together, we can create and save a vector of file names:</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb336-1" title="1">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb336-2" title="2">filenamez &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb336-3" title="3"></a>
<a class="sourceLine" id="cb336-4" title="4"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb336-5" title="5">  filenamez[i] &lt;-<span class="st"> </span>first_five[i]</a>
<a class="sourceLine" id="cb336-6" title="6">}</a>
<a class="sourceLine" id="cb336-7" title="7"></a>
<a class="sourceLine" id="cb336-8" title="8">filenamez</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot; &quot;viral_proteins_001.tsv&quot; &quot;viral_proteins_002.tsv&quot;
## [4] &quot;viral_proteins_003.tsv&quot; &quot;viral_proteins_004.tsv&quot;</code></pre>
<hr />
<p>And finally calculate the length of each protein:</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb338-1" title="1">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb338-2" title="2">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb338-3" title="3">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb338-4" title="4"></a>
<a class="sourceLine" id="cb338-5" title="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb338-6" title="6">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb338-7" title="7">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb338-8" title="8">    )</a>
<a class="sourceLine" id="cb338-9" title="9">}</a>
<a class="sourceLine" id="cb338-10" title="10"></a>
<a class="sourceLine" id="cb338-11" title="11">lengthz</a></code></pre></div>
<pre><code>## [1] &quot;1248&quot; &quot;1254&quot; &quot;177&quot;  &quot;137&quot;  &quot;117&quot;</code></pre>
<hr />
<p>Now we can take off the <code>[1:5]</code> limiter, and do the whole set:</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb340-1" title="1">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb340-2" title="2">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb340-3" title="3">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb340-4" title="4"></a>
<a class="sourceLine" id="cb340-5" title="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb340-6" title="6">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb340-7" title="7">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb340-8" title="8">    )</a>
<a class="sourceLine" id="cb340-9" title="9">}</a></code></pre></div>
<hr />
<p>How do we go about visualizing/analyzing this?</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb341-1" title="1"><span class="kw">hist</span>(lengthz)</a></code></pre></div>
<pre><code>## Error in hist.default(lengthz): &#39;x&#39; must be numeric</code></pre>
<p>er what…? Debug! What is <code>lengthz</code>?</p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb343-1" title="1"><span class="kw">str</span>(lengthz)</a></code></pre></div>
<pre><code>##  chr [1:244] &quot;1248&quot; &quot;1254&quot; &quot;177&quot; &quot;137&quot; &quot;117&quot; &quot;117&quot; &quot;858&quot; &quot;1242&quot; &quot;1255&quot; ...</code></pre>
<p>Character? Let’s try numeric instead…</p>
<hr />
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb345-1" title="1">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb345-2" title="2">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb345-3" title="3">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;numeric&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb345-4" title="4"></a>
<a class="sourceLine" id="cb345-5" title="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb345-6" title="6">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb345-7" title="7">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb345-8" title="8">    )</a>
<a class="sourceLine" id="cb345-9" title="9">}</a></code></pre></div>
<hr />
<p>Base R histogram</p>
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb346-1" title="1"><span class="kw">hist</span>(lengthz,<span class="dt">breaks=</span><span class="dv">50</span>)</a></code></pre></div>
<p><img src="515-loops_files/figure-html/base_viz_protein_lengths_fix-1.png" width="672" /></p>
<hr />
<p>More fancy ggplot</p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb347-1" title="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb347-2" title="2"><span class="kw">ggplot</span>( <span class="kw">data.frame</span>(<span class="dt">length=</span>lengthz) )<span class="op">+</span><span class="kw">theme_classic</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb347-3" title="3"><span class="st">  </span><span class="kw">aes</span>(<span class="dt">x=</span>length)<span class="op">+</span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">50</span>)</a></code></pre></div>
<p><img src="515-loops_files/figure-html/ggplot_viz_protein_lengths-1.png" width="672" /></p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb348-1" title="1"><span class="kw">summary</span>(lengthz)</a></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     0.0   230.0   626.0   726.3  1088.5  3390.0</code></pre>
<hr />
<p>And that’s about how long the viral structural proteins in this database
tend to be.</p>
<ul>
<li>How do we extend this analysis to add more files?</li>
<li>How do we calculate new properties about the sequences?</li>
<li>How do we store that?</li>
</ul>

</div>
</div>
<div id="extending-the-workflow" class="section level3">
<h3><span class="header-section-number">4.1.3</span> Extending the workflow</h3>
<p>How do you think about/plan the workflow we’ve built?</p>
<p>One way is to flatten out all the tasks,
and to script each individual task every time it is done.
This requires the author, user, and reader to understand a lot of complexity.</p>
<div class="incremental">
<div class="mermaid">
graph TD
    classDef default line-height:12px;
    classDef three line-height:12px,fill:yellow;
    you[analyst] --> B[finding files] &    C[making a vector] &    D[looping] &    E[reading files] &    F[getting protein seq] &    G[calculting and saving nchar] &    H[plotting nchar]
    B:::three;
    C:::three;
    D:::three;
    E:::three;
    F:::three;
    G:::three;
    H:::three;
</div>
<p>Another way to approach this is to cluster them into a hierarchy of modules.</p>
<div class="mermaid">
graph TD
    classDef default line-height:12px;
    classDef one  line-height:12px,fill:pink;
    classDef two line-height:12px,fill:cyan;
    classDef three line-height:12px,fill:yellow;
    you2[analyst] --> find & read & plot;
    find:::one --> B;
    B[finding files]:::three;
    read:::one --> looping & reading;
    looping:::two --> C & D;
    C[making a vector]:::three;
    D[looping]:::three;
    reading[reading and calc]:::two --> E & F & G;
    E[reading files]:::three;
    F[getting protein seq]:::three;
    G[calculting and saving nchar]:::three;
    plot:::one --> H;
    H[plotting nchar]:::three;
</div>
<p>In this organization, the analyst can operate at levels of
<span style="color:magenta">steps</span>,
<span style="color:cyan">modules</span>, and
<span style="color:gold">specific instructions</span>,
depending on what is needed.</p>
</div>
<hr />
<p>Organizing your workflows into composable modules lets you extend these to
un-ancipiated complexity. You could imaging using these steps or modules:</p>
<div class="mermaid">
graph LR
    classDef default line-height:12px;
    vp[read viral proteins] --> cl[calculate lengths] --> hist[histogram]
</div>
<p>in new ways by composing the elements together, to analyze a different source
of proteins, with a new analysis, with similar plots:</p>
<div class="mermaid">
graph LR
    classDef default line-height:12px;
    vp[read viral proteins] --> cl[calculate lengths] --> hist[histogram]
    hp[read human proteins] --> cl --> boxplot[boxplot]
    vp --> cmotif[find a particular motif] --> motifhist[histogram]
    hp --> cmotif --> motifboxplot[boxplot]
</div>
<p>Where maybe these are how the inputs/outputs are defined:</p>
<pre><code>reading_proteins:
    input: directory path of proteins to read
    output: protein sequences
calculating lengths:
    input: protein sequences
    output: numeric vector, of lengths
finding a motif:
    input: protein sequences
    output: numeric vector, of presence/absence</code></pre>
<hr />
<div id="looking-for-a-smoking-motif" class="section level4">
<h4><span class="header-section-number">4.1.3.1</span> Looking for a smoking motif…</h4>
<p>Some folks believe the insertion of an amino-acid sequence of <code>PRRA</code>, relative
to other related coronaviruses, is very rare and a clear sign of engineering.
I don’t know hardly anything about furin-cleavage sites or viral evolution,
but we can use the tools we have to look for this sequence in the files.</p>
<p>Let’s do that<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>:</p>
<ol style="list-style-type: decimal">
<li><p>Figure out how to search for “PRRA” in a character string of a protein
sequence.</p>
<ul>
<li>try using <code>grepl</code>, like so: <code>grepl("PRRA",protein_sequence)</code>
where <code>protein_sequence</code> is the variable containing the
protein sequence (similar to what you put into <code>nchar</code> before!).</li>
</ul></li>
<li><p>Modify your <code>for</code> loop to now look for this motif instead of calculating
number of characters.
Save the values in a “numeric” vector.
<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a></p></li>
<li><p>Plot the results. How often do you find this motif in this dataset?</p></li>
</ol>
<hr />
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>
mermaid.initialize({
    startOnLoad: false, // don't render initially, instead....
    logLevel: 3, // for debugging
});
function initMermaid(s) { 
  var diagrams = document.querySelectorAll('.mermaid');
  var i;
  for(i=0;i<diagrams.length;i++){
    if(diagrams[i].offsetWidth>0){
      mermaid.init(undefined, diagrams[i]);
    }
  }
}
initMermaid();
</script>

</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="3">
<li id="fn3"><p>One crucial aspect missing is the actual bash commands used. I did not
save them as a script, and I <em>should</em> go back, save them in a script,
and re-run that to make sure I get the same result!<a href="analyzing-viral-protein-sequences.html#fnref3" class="footnote-back">↩</a></p></li>
<li id="fn4"><p>Often this is typed into computers and scripts as ", as the tab key
will often not put a TAB character into where you’re typing.<a href="analyzing-viral-protein-sequences.html#fnref4" class="footnote-back">↩</a></p></li>
<li id="fn5"><p>
…since we’re just demonstrating R coding and not representing ourselves as
being familiar enough with the field to properly interpret these results.
I have no idea what occurance of this would mean or not mean, or are we
doing any sort of statistics or comparisons to null expectations.<a href="analyzing-viral-protein-sequences.html#fnref5" class="footnote-back">↩</a></p></li>
<li id="fn6"><p>Why not logical, since these are logical results? It makes the <code>hist()</code>
step slightly easier.<a href="analyzing-viral-protein-sequences.html#fnref6" class="footnote-back">↩</a></p></li>
</ol>
</div>
<p style="margin-top:10em;text-align:center;font-size:small;">
    <a href="https://github.com/darachm/dll-r/blob/main/LICENSE.md">
        Licensed Creative Commons Attribution-NonCommercial-ShareAlike 4.0
    </a>
</p>
<p style="text-align:center;font-size:small;">
    <a href="https://github.com/rstudio/bookdown" target="blank">
        Website typeset with bookdown
    </a>
</p>

<script> 
// This is to look through, everything that's a floating lone double dash,
// delete it. This is left over for stuff that's trying to be pauses for
// the slides
var listz = document.getElementsByTagName("p");
for (var i=0; i < listz.length; i++) { 
    if (listz[i].innerText === "–") { listz[i].innerText = "" } 
} 
</script>
            </section>

          </div>
        </div>
      </div>
<a href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="troubleshooting-getting-help-dissecting-problems.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/darachm/dll-r/edit/main/510-viral-protein-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["DLL 2021 R syllabus site.html"],
"toc": {
"collapse": "subsection"
},
"highlight": "zenburn"
});
});
</script>

</body>

</html>
