<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.3 Packages - install and use | DLL 2021, R section</title>
  <meta name="description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="4.3 Packages - install and use | DLL 2021, R section" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="github-repo" content="darachm/dll-r" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.3 Packages - install and use | DLL 2021, R section" />
  
  <meta name="twitter:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  

<meta name="author" content="mesako" />
<meta name="author" content="Margaret" />
<meta name="author" content="Samson" />
<meta name="author" content="Zac" />
<meta name="author" content="darachm" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="analyzing-viral-protein-sequences.html"/>
<link rel="next" href="getting-help-troubleshooting.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DLL 2021 R workshops</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Workshop Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="workshop-goals.html"><a href="workshop-goals.html"><i class="fa fa-check"></i><b>1.1</b> Workshop goals</a></li>
<li class="chapter" data-level="1.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html"><i class="fa fa-check"></i><b>1.2</b> Structure and resources</a><ul>
<li class="chapter" data-level="1.2.1" data-path="structure-and-resources.html"><a href="structure-and-resources.html#workshop-schedule"><i class="fa fa-check"></i><b>1.2.1</b> Workshop schedule</a></li>
<li class="chapter" data-level="1.2.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html#asynchronous-sessions"><i class="fa fa-check"></i><b>1.2.2</b> Asynchronous sessions</a></li>
<li class="chapter" data-level="1.2.3" data-path="structure-and-resources.html"><a href="structure-and-resources.html#synchronous-sessions"><i class="fa fa-check"></i><b>1.2.3</b> Synchronous sessions</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="workshop-expectations.html"><a href="workshop-expectations.html"><i class="fa fa-check"></i><b>1.3</b> Workshop expectations</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="day-3-intoduction-to-r.html"><a href="day-3-intoduction-to-r.html"><i class="fa fa-check"></i><b>2</b> Day 3 - intoduction to R</a><ul>
<li class="chapter" data-level="2.1" data-path="using-rstudio.html"><a href="using-rstudio.html"><i class="fa fa-check"></i><b>2.1</b> Using RStudio</a><ul>
<li class="chapter" data-level="2.1.1" data-path="using-rstudio.html"><a href="using-rstudio.html#r-scripts-in-rstudio"><i class="fa fa-check"></i><b>2.1.1</b> R Scripts in RStudio</a></li>
<li class="chapter" data-level="2.1.2" data-path="using-rstudio.html"><a href="using-rstudio.html#console-and-terminal"><i class="fa fa-check"></i><b>2.1.2</b> Console and Terminal</a></li>
<li class="chapter" data-level="2.1.3" data-path="using-rstudio.html"><a href="using-rstudio.html#environment-and-history"><i class="fa fa-check"></i><b>2.1.3</b> Environment and History</a></li>
<li class="chapter" data-level="2.1.4" data-path="using-rstudio.html"><a href="using-rstudio.html#files-plots-and-packages"><i class="fa fa-check"></i><b>2.1.4</b> Files, Plots, and Packages</a></li>
<li class="chapter" data-level="2.1.5" data-path="using-rstudio.html"><a href="using-rstudio.html#putting-it-all-together"><i class="fa fa-check"></i><b>2.1.5</b> Putting It All Together</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-rmds.html"><a href="using-rmds.html"><i class="fa fa-check"></i><b>2.2</b> Using Rmds</a><ul>
<li class="chapter" data-level="2.2.1" data-path="using-rmds.html"><a href="using-rmds.html#basic-structure"><i class="fa fa-check"></i><b>2.2.1</b> Basic Structure</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-rmds.html"><a href="using-rmds.html#overview-of-markdown"><i class="fa fa-check"></i><b>2.2.2</b> Overview of Markdown</a></li>
<li class="chapter" data-level="2.2.3" data-path="using-rmds.html"><a href="using-rmds.html#overview-of-how-to-write-an-r-chunk"><i class="fa fa-check"></i><b>2.2.3</b> Overview of how to write an R chunk</a></li>
<li class="chapter" data-level="2.2.4" data-path="using-rmds.html"><a href="using-rmds.html#final-thoughts"><i class="fa fa-check"></i><b>2.2.4</b> Final Thoughts</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Getting started in R</a><ul>
<li class="chapter" data-level="2.3.1" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#getting-help-in-r"><i class="fa fa-check"></i><b>2.3.1</b> Getting help in R</a></li>
<li class="chapter" data-level="2.3.2" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-variables-and-data-types"><i class="fa fa-check"></i><b>2.3.2</b> Using variables and data types</a></li>
<li class="chapter" data-level="2.3.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-functions"><i class="fa fa-check"></i><b>2.3.3</b> Using functions</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html"><i class="fa fa-check"></i><b>2.4</b> Working with vectors</a><ul>
<li class="chapter" data-level="2.4.1" data-path="working-with-vectors.html"><a href="working-with-vectors.html#building-vectors"><i class="fa fa-check"></i><b>2.4.1</b> Building vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="working-with-vectors.html"><a href="working-with-vectors.html#indexing-and-subsetting-vectors"><i class="fa fa-check"></i><b>2.4.2</b> Indexing and subsetting vectors</a></li>
<li class="chapter" data-level="2.4.3" data-path="working-with-vectors.html"><a href="working-with-vectors.html#using-logic-and-logicals"><i class="fa fa-check"></i><b>2.4.3</b> Using logic and logicals</a></li>
<li class="chapter" data-level="2.4.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html#modifying-vectors"><i class="fa fa-check"></i><b>2.4.4</b> Modifying vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html"><i class="fa fa-check"></i><b>2.5</b> Working with dataframes</a><ul>
<li class="chapter" data-level="2.5.1" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#indexing-and-subsetting-dataframes"><i class="fa fa-check"></i><b>2.5.1</b> Indexing and subsetting dataframes</a></li>
<li class="chapter" data-level="2.5.2" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#exploring-dataframes"><i class="fa fa-check"></i><b>2.5.2</b> Exploring dataframes</a></li>
<li class="chapter" data-level="2.5.3" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#building-and-modifying-dataframes"><i class="fa fa-check"></i><b>2.5.3</b> Building and modifying dataframes</a></li>
<li class="chapter" data-level="2.5.4" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#handling-datatypes-in-dataframes"><i class="fa fa-check"></i><b>2.5.4</b> Handling datatypes in dataframes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="day-4-tidyverse-and-visualizations.html"><a href="day-4-tidyverse-and-visualizations.html"><i class="fa fa-check"></i><b>3</b> Day 4 - tidyverse and visualizations</a><ul>
<li class="chapter" data-level="3.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html"><i class="fa fa-check"></i><b>3.1</b> Reading and processing data</a><ul>
<li class="chapter" data-level="3.1.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#loading-a-file-into-r"><i class="fa fa-check"></i><b>3.1.1</b> Loading a file into R</a></li>
<li class="chapter" data-level="3.1.2" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#handling-issues-in-data"><i class="fa fa-check"></i><b>3.1.2</b> Handling issues in data</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html"><i class="fa fa-check"></i><b>3.2</b> Tidyverse 1/2: functions &amp; pipes</a><ul>
<li class="chapter" data-level="3.2.1" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#welcome-to-the-tidyverse"><i class="fa fa-check"></i><b>3.2.1</b> Welcome to the tidyverse!</a></li>
<li class="chapter" data-level="3.2.2" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#load-tidyverse"><i class="fa fa-check"></i><b>3.2.2</b> Load tidyverse</a></li>
<li class="chapter" data-level="3.2.3" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#how-to-pipe"><i class="fa fa-check"></i><b>3.2.3</b> How to pipe %&gt;%</a></li>
<li class="chapter" data-level="3.2.4" data-path="tidyverse-12-functions-pipes.html"><a href="tidyverse-12-functions-pipes.html#essential-tidyverse-functions"><i class="fa fa-check"></i><b>3.2.4</b> Essential tidyverse functions</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html"><i class="fa fa-check"></i><b>3.3</b> Tidyverse 2/2: reshaping data</a><ul>
<li class="chapter" data-level="3.3.1" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html#why-tidy-data"><i class="fa fa-check"></i><b>3.3.1</b> Why tidy data</a></li>
<li class="chapter" data-level="3.3.2" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html#how-to-tidy-data"><i class="fa fa-check"></i><b>3.3.2</b> How to tidy data</a></li>
<li class="chapter" data-level="3.3.3" data-path="tidyverse-22-reshaping-data.html"><a href="tidyverse-22-reshaping-data.html#data-exploration-with-tidyverse"><i class="fa fa-check"></i><b>3.3.3</b> Data exploration with tidyverse</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html"><i class="fa fa-check"></i><b>3.4</b> Making plots with ggplot2</a><ul>
<li class="chapter" data-level="3.4.1" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#getting-started-with-a-ggplot"><i class="fa fa-check"></i><b>3.4.1</b> Getting started with a ggplot</a></li>
<li class="chapter" data-level="3.4.2" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#changing-plot-aesthetics"><i class="fa fa-check"></i><b>3.4.2</b> Changing plot aesthetics</a></li>
<li class="chapter" data-level="3.4.3" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#exploring-simple-plots"><i class="fa fa-check"></i><b>3.4.3</b> Exploring simple plots</a></li>
<li class="chapter" data-level="3.4.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#visualizing-between-groups"><i class="fa fa-check"></i><b>3.4.4</b> Visualizing between groups</a></li>
<li class="chapter" data-level="3.4.5" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#generating-heatmaps"><i class="fa fa-check"></i><b>3.4.5</b> Generating heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html"><i class="fa fa-check"></i><b>3.5</b> Making scientific figures</a><ul>
<li class="chapter" data-level="3.5.1" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#plotting-error-bars"><i class="fa fa-check"></i><b>3.5.1</b> Plotting error bars</a></li>
<li class="chapter" data-level="3.5.2" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#showing-trends-in-data"><i class="fa fa-check"></i><b>3.5.2</b> Showing trends in data</a></li>
<li class="chapter" data-level="3.5.3" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#saving-figures-locally"><i class="fa fa-check"></i><b>3.5.3</b> Saving figures locally</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html"><i class="fa fa-check"></i><b>3.6</b> Applying basic stats</a><ul>
<li class="chapter" data-level="3.6.1" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#describing-quantitative-variables"><i class="fa fa-check"></i><b>3.6.1</b> Describing quantitative variables</a></li>
<li class="chapter" data-level="3.6.2" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#finding-differences-in-data"><i class="fa fa-check"></i><b>3.6.2</b> Finding differences in data</a></li>
<li class="chapter" data-level="3.6.3" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#identifying-correlations"><i class="fa fa-check"></i><b>3.6.3</b> Identifying correlations</a></li>
<li class="chapter" data-level="3.6.4" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#producing-linear-models"><i class="fa fa-check"></i><b>3.6.4</b> Producing linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="day-5-workflows-loops-packages-and-help.html"><a href="day-5-workflows-loops-packages-and-help.html"><i class="fa fa-check"></i><b>4</b> Day 5 - workflows, loops, packages, and help</a><ul>
<li class="chapter" data-level="4.1" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html"><i class="fa fa-check"></i><b>4.1</b> What is a workflow?</a><ul>
<li class="chapter" data-level="4.1.1" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html#organizing-files"><i class="fa fa-check"></i><b>4.1.1</b> Organizing files</a></li>
<li class="chapter" data-level="4.1.2" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html#use-a-notebook-to-generate-reports"><i class="fa fa-check"></i><b>4.1.2</b> Use a notebook to generate reports</a></li>
<li class="chapter" data-level="4.1.3" data-path="what-is-a-workflow.html"><a href="what-is-a-workflow.html#sharing-with-others"><i class="fa fa-check"></i><b>4.1.3</b> Sharing with others</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html"><i class="fa fa-check"></i><b>4.2</b> Analyzing viral protein sequences</a><ul>
<li class="chapter" data-level="4.2.1" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#reading-in-tab-delimited-file"><i class="fa fa-check"></i><b>4.2.1</b> Reading in tab-delimited file</a></li>
<li class="chapter" data-level="4.2.2" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#string-stuff"><i class="fa fa-check"></i><b>4.2.2</b> String stuff</a></li>
<li class="chapter" data-level="4.2.3" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#loops---doing-a-similar-task-multiple-times"><i class="fa fa-check"></i><b>4.2.3</b> Loops - doing a similar task multiple times</a></li>
<li class="chapter" data-level="4.2.4" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#storing-values-from-a-loop"><i class="fa fa-check"></i><b>4.2.4</b> Storing values from a loop</a></li>
<li class="chapter" data-level="4.2.5" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#extending-the-workflow"><i class="fa fa-check"></i><b>4.2.5</b> Extending the workflow</a></li>
<li class="chapter" data-level="4.2.6" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#conclusion"><i class="fa fa-check"></i><b>4.2.6</b> Conclusion</a></li>
<li class="chapter" data-level="4.2.7" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#totally-optional-supplements-for-your-edification"><i class="fa fa-check"></i><b>4.2.7</b> Totally optional supplements, for your edification</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html"><i class="fa fa-check"></i><b>4.3</b> Packages - install and use</a><ul>
<li class="chapter" data-level="4.3.1" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#cran-is-the-canonical-r-package-resource"><i class="fa fa-check"></i><b>4.3.1</b> CRAN is the canonical R package resource</a></li>
<li class="chapter" data-level="4.3.2" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#using-packages-without-library-ing-them"><i class="fa fa-check"></i><b>4.3.2</b> Using packages without <code>library()</code>-ing them</a></li>
<li class="chapter" data-level="4.3.3" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#github-is-a-common-place-for-sharing-packges-in-development"><i class="fa fa-check"></i><b>4.3.3</b> Github is a common place for sharing packges in development</a></li>
<li class="chapter" data-level="4.3.4" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#bioconductor-is-a-great-place-for-bio-related-packages"><i class="fa fa-check"></i><b>4.3.4</b> Bioconductor is a great place for bio-related packages</a></li>
<li class="chapter" data-level="4.3.5" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#what-packages-do-you-have-already"><i class="fa fa-check"></i><b>4.3.5</b> What packages do you have already?</a></li>
<li class="chapter" data-level="4.3.6" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#major-problems-with-package-installations"><i class="fa fa-check"></i><b>4.3.6</b> Major problems with package installations</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html"><i class="fa fa-check"></i><b>4.4</b> Getting help - troubleshooting</a><ul>
<li class="chapter" data-level="4.4.1" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#where-to-start-general-help"><i class="fa fa-check"></i><b>4.4.1</b> Where to start? General help</a></li>
<li class="chapter" data-level="4.4.2" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#debugging-specific-problems"><i class="fa fa-check"></i><b>4.4.2</b> Debugging specific problems</a></li>
<li class="chapter" data-level="4.4.3" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#minimal-reproducible-examples"><i class="fa fa-check"></i><b>4.4.3</b> Minimal reproducible examples</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="the-struggle.html"><a href="the-struggle.html"><i class="fa fa-check"></i><b>4.5</b> The Struggle</a></li>
<li class="chapter" data-level="4.6" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>4.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="appendix.html"><a href="appendix.html"><i class="fa fa-check"></i><b>5</b> Appendix</a><ul>
<li class="chapter" data-level="5.1" data-path="some-links-and-some-extras.html"><a href="some-links-and-some-extras.html"><i class="fa fa-check"></i><b>5.1</b> Some links and some extras</a><ul>
<li class="chapter" data-level="5.1.1" data-path="some-links-and-some-extras.html"><a href="some-links-and-some-extras.html#ideas-for-figures"><i class="fa fa-check"></i><b>5.1.1</b> Ideas for figures</a></li>
<li class="chapter" data-level="5.1.2" data-path="some-links-and-some-extras.html"><a href="some-links-and-some-extras.html#more-links"><i class="fa fa-check"></i><b>5.1.2</b> More links</a></li>
<li class="chapter" data-level="5.1.3" data-path="some-links-and-some-extras.html"><a href="some-links-and-some-extras.html#more-about-workflows"><i class="fa fa-check"></i><b>5.1.3</b> More about workflows</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html"><i class="fa fa-check"></i><b>5.2</b> Style - writing (re)usable code</a><ul>
<li class="chapter" data-level="5.2.1" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#functions-as-modular-steps"><i class="fa fa-check"></i><b>5.2.1</b> Functions as modular steps</a></li>
<li class="chapter" data-level="5.2.2" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#tips-for-modular-workflows"><i class="fa fa-check"></i><b>5.2.2</b> Tips for modular workflows</a></li>
<li class="chapter" data-level="5.2.3" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#apply-is-a-popular-tool"><i class="fa fa-check"></i><b>5.2.3</b> Apply is a popular tool</a></li>
<li class="chapter" data-level="5.2.4" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#learn-from-examples"><i class="fa fa-check"></i><b>5.2.4</b> Learn from examples</a></li>
<li class="chapter" data-level="5.2.5" data-path="style-writing-reusable-code.html"><a href="style-writing-reusable-code.html#style-guides-can-be-inspiring"><i class="fa fa-check"></i><b>5.2.5</b> Style guides can be inspiring</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DLL 2021, R section</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="packages---install-and-use" class="section level2">
<h2><span class="header-section-number">4.3</span> Packages - install and use</h2>
<p>Think about sharing your code from the last few days.
How would you share this with the other people in the course?
Can they use it easily and flexibly to handle lots of different problems?</p>
<p>R, like other languages, is built on a package system - if you wrap up your
code in particular expected ways and put it in particular expected places,
it is very easy for others to get and use your code</p>
<p>Packages are
<a href="https://ourcodingclub.github.io/tutorials/writing-r-package/">said to be easy to make</a>,
<a href="https://hilaryparker.com/2014/04/29/writing-an-r-package-from-scratch/">from scratch</a>.
There’s more <a href="https://r-pkgs.org/">comprehensive instruction here</a>.
Personally, the author has never had cause to make one, but I can report that
they’re very easy to use!</p>
<hr />
<div id="cran-is-the-canonical-r-package-resource" class="section level3">
<h3><span class="header-section-number">4.3.1</span> CRAN is the canonical R package resource</h3>
<p>The Comprehensive R Archive Netowrk
<a href="https://cran.r-project.org/web/packages/">CRAN</a>
is an integral part of the R environment.
This is a global and free resource for storing and distributing code
as R packages - well documented and designed chunks of code.</p>
<p>You can install packages using the RStudio IDE, but it’s most important
to know how to do it in R - what’s happening just under the hood.</p>
<p>Use <code>library</code> to load packages.
You should be able to load a package you’ve
already installed.
Call <code>library()</code> function, with an argument of the library name,
as a string.</p>
<p>For example:</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb428-1" title="1"><span class="kw">library</span>(<span class="st">&quot;stringdist&quot;</span>)</a>
<a class="sourceLine" id="cb428-2" title="2"><span class="kw">library</span>(stringdist)</a></code></pre></div>
<p>Don’t have it? You can look for packages by using a search engine, or finding
suggestions in the documentation, papers. Once you have a package name, use</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb429-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;stringdist&quot;</span>)</a></code></pre></div>
<p>to install it. You may need to select a “mirror”. This is just where you
are downloading from, usually just go with “0-Cloud”
<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a>.</p>
<p>What did this do?
It retrieves and downloads the package’s code and documentation to
what’s called your library.
That’s a folder on your computer where the compiled versions of the packages
are located, ready to be loaded when you want to use them.</p>
<p>Try it out.</p>
<p>Want to try it again? You can uninstall packages by removing them:</p>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb430-1" title="1"><span class="kw">remove.packages</span>(<span class="st">&quot;stringdist&quot;</span>)</a></code></pre></div>
<p>Each package has a <a href="https://cran.r-project.org/web/packages/stringdist/index.html">page on CRAN</a>
with lots more information. The reference manual is a very specific reference
document with per function level documentation.
Vingettes are write ups that are more explaining how to do things - usually.</p>
<div id="ggrepel" class="section level4">
<h4><span class="header-section-number">4.3.1.1</span> ggrepel</h4>
<p>Sometimes points are way to close, and you can’t label them properly.
For example,</p>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb431-1" title="1"><span class="kw">library</span>(tidyverse)</a></code></pre></div>
<pre><code>## ── Attaching packages ───────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>## ✔ ggplot2 3.3.2     ✔ purrr   0.3.4
## ✔ tibble  3.0.4     ✔ dplyr   1.0.2
## ✔ tidyr   1.1.2     ✔ stringr 1.4.0
## ✔ readr   1.4.0     ✔ forcats 0.5.0</code></pre>
<pre><code>## ── Conflicts ──────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb435-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1234</span>) <span class="co"># this is just to get the same results each time we run</span></a>
<a class="sourceLine" id="cb435-2" title="2"></a>
<a class="sourceLine" id="cb435-3" title="3"><span class="kw">read_csv</span>(<span class="st">&quot;data/iris.csv&quot;</span>,<span class="dt">col_types=</span><span class="st">&#39;nnnncc&#39;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb435-4" title="4"><span class="st">    </span><span class="kw">slice_sample</span>(<span class="dt">n=</span><span class="dv">20</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb435-5" title="5"><span class="st">    </span><span class="kw">ggplot</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb435-6" title="6"><span class="st">    </span><span class="kw">aes</span>(<span class="dt">x=</span>Sepal.Length,<span class="dt">y=</span>Sepal.Width,<span class="dt">col=</span>Species,<span class="dt">label=</span>Species)<span class="op">+</span></a>
<a class="sourceLine" id="cb435-7" title="7"><span class="st">    </span><span class="kw">geom_label</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb435-8" title="8"><span class="st">    </span><span class="kw">geom_point</span>()</a></code></pre></div>
<pre><code>## Warning: 4 parsing failures.
## row          col expected actual            file
##  44 Sepal.Length a number    n/a &#39;data/iris.csv&#39;
##  44 Sepal.Width  a number    n/a &#39;data/iris.csv&#39;
## 148 Petal.Length a number    n/a &#39;data/iris.csv&#39;
## 148 Petal.Width  a number    n/a &#39;data/iris.csv&#39;</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_label).</code></pre>
<p><img src="550-packages_files/figure-html/iris_overplot_example-1.png" width="672" /></p>
<p>The <code>ggrepel</code> library, from CRAN, is a handy way to achieve this:</p>
<div class="sourceCode" id="cb438"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb438-1" title="1"><span class="kw">library</span>(ggrepel)</a>
<a class="sourceLine" id="cb438-2" title="2"></a>
<a class="sourceLine" id="cb438-3" title="3"><span class="kw">set.seed</span>(<span class="dv">1234</span>) <span class="co"># this is just to get the same results each time we run</span></a>
<a class="sourceLine" id="cb438-4" title="4"></a>
<a class="sourceLine" id="cb438-5" title="5"><span class="kw">read_csv</span>(<span class="st">&quot;data/iris.csv&quot;</span>,<span class="dt">col_types=</span><span class="st">&#39;nnnncc&#39;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb438-6" title="6"><span class="st">    </span><span class="kw">slice_sample</span>(<span class="dt">n=</span><span class="dv">20</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb438-7" title="7"><span class="st">    </span><span class="kw">ggplot</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb438-8" title="8"><span class="st">    </span><span class="kw">aes</span>(<span class="dt">x=</span>Sepal.Length,<span class="dt">y=</span>Sepal.Width,<span class="dt">col=</span>Species,<span class="dt">label=</span>Species)<span class="op">+</span></a>
<a class="sourceLine" id="cb438-9" title="9"><span class="st">    </span><span class="kw">geom_label_repel</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb438-10" title="10"><span class="st">    </span><span class="kw">geom_point</span>()</a></code></pre></div>
<pre><code>## Warning: 4 parsing failures.
## row          col expected actual            file
##  44 Sepal.Length a number    n/a &#39;data/iris.csv&#39;
##  44 Sepal.Width  a number    n/a &#39;data/iris.csv&#39;
## 148 Petal.Length a number    n/a &#39;data/iris.csv&#39;
## 148 Petal.Width  a number    n/a &#39;data/iris.csv&#39;</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_label_repel).</code></pre>
<p><img src="550-packages_files/figure-html/iris_ggrepel-1.png" width="672" /></p>
<hr />
</div>
</div>
<div id="using-packages-without-library-ing-them" class="section level3">
<h3><span class="header-section-number">4.3.2</span> Using packages without <code>library()</code>-ing them</h3>
<p>Some packages you don’t really want to have to go through the trouble of
loading, you’ll just want one function.
For example, this website is generated with one function, <code>render_book()</code>
in the package <code>bookdown</code>. Instead of:</p>
<div class="sourceCode" id="cb441"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb441-1" title="1"><span class="kw">library</span>(bookdown)</a>
<a class="sourceLine" id="cb441-2" title="2"><span class="kw">render_book</span>()</a></code></pre></div>
<p>I can save a line by just using <code>::</code></p>
<div class="sourceCode" id="cb442"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb442-1" title="1">bookdown<span class="op">::</span><span class="kw">render_book</span>()</a></code></pre></div>
<p>This means I am looking in library <code>bookdown</code> for function <code>render_book</code>.
The function is hidden again right afterwards, but it is useful if you just
want to run one function from a library, once.</p>
<p>This is <em>super handy</em> when wanting to typeset (<code>render</code>) a <code>Rmd</code> file:</p>
<div class="sourceCode" id="cb443"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb443-1" title="1">rmarkdown<span class="op">::</span><span class="kw">render</span>()</a></code></pre></div>
<p>I <em>never</em> load the <code>rmarkdown</code> package, I just type the above. A lot.
<a href="#fn8" class="footnote-ref" id="fnref8"><sup>8</sup></a></p>
<hr />
</div>
<div id="github-is-a-common-place-for-sharing-packges-in-development" class="section level3">
<h3><span class="header-section-number">4.3.3</span> Github is a common place for sharing packges in development</h3>
<p>Installing from github requires some different programming -
and there’s libraries with functions for this.</p>
<div class="sourceCode" id="cb444"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb444-1" title="1"><span class="kw">library</span>(<span class="st">&quot;remotes&quot;</span>)</a></code></pre></div>
<p>This used to be part of the <code>devtools</code> package, but the authors have been
“uncoupling” these to make each package simple and composable.</p>
<div id="lets-install-a-package-off-of-github." class="section level4">
<h4><span class="header-section-number">4.3.3.1</span> Let’s install a package off of github.</h4>
<p>Here’s a cool <a href="https://github.com/ewenme/ghibli">Ghibli color palette</a>,
so you can change sets of colors for your plots.
Try installing it with:</p>
<div class="sourceCode" id="cb445"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb445-1" title="1">remotes<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;ewenme/ghibli&quot;</span>)</a></code></pre></div>
<p>Then adapt one of your previous plots (that has colors) to use these new
color palettes.</p>
</div>
<div id="lets-install-another-one-to-check-the-previous" class="section level4">
<h4><span class="header-section-number">4.3.3.2</span> Let’s install another one, to check the previous</h4>
<p>Colorblind folks can’t distinguish between certain colors so well.
Designing figures so that colorblind folks can readily distinguish the
relevant differences is a <em>good idea</em>. It makes for a better figure, just
like increasing the font size and cleaning up distracting artifacts.</p>
<p>The <code>ghibli</code> package above has been claimed to be somewhat good for
plotting for universal visibility. Is that true?</p>
<p>Let’s install this package to check:</p>
<div class="sourceCode" id="cb446"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb446-1" title="1">remotes<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;nowosad/colorblindcheck&quot;</span>,<span class="dt">upgrade=</span><span class="st">&quot;never&quot;</span>)</a></code></pre></div>
<p>Then</p>
<div class="sourceCode" id="cb447"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb447-1" title="1"><span class="kw">library</span>(ghibli)</a>
<a class="sourceLine" id="cb447-2" title="2"><span class="kw">library</span>(colorblindcheck)</a>
<a class="sourceLine" id="cb447-3" title="3"></a>
<a class="sourceLine" id="cb447-4" title="4"><span class="kw">palette_check</span>(ghibli_palettes<span class="op">$</span>PonyoMedium,<span class="dt">plot=</span>T)</a></code></pre></div>
<p><img src="550-packages_files/figure-html/checkitout-1.png" width="672" /></p>
<pre><code>##           name n tolerance ncp ndcp min_dist mean_dist max_dist
## 1       normal 7  9.939857  21   21 9.939857  36.11891 59.92427
## 2 deuteranopia 7  9.939857  21   19 7.044647  30.14324 52.02502
## 3   protanopia 7  9.939857  21   20 8.068943  26.94418 50.93756
## 4   tritanopia 7  9.939857  21   20 4.587519  35.34529 60.29190</code></pre>
<p>Other packages on there:</p>
<ul>
<li><a href="https://github.com/rossdrucker/sportyR">Sports field plotting</a></li>
<li><a href="https://github.com/wilkelab/ggridges">Ridgeplots in ggplot2</a></li>
<li>what else?</li>
</ul>
<!-- remotes::install_github("wilkelab/ggridges" -->
<hr />
</div>
</div>
<div id="bioconductor-is-a-great-place-for-bio-related-packages" class="section level3">
<h3><span class="header-section-number">4.3.4</span> Bioconductor is a great place for bio-related packages</h3>
<p>Bioconductor is a giant repository of just scientific, well bio-focused,
packages. It’s where a lot of bioinformatics software gets published.
However, it has it’s own package system and thus package manager.
You will have to install it … from CRAN. Like so:</p>
<div class="sourceCode" id="cb449"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb449-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a></code></pre></div>
<p>That installs a package with a function that then installs Bioconductor
packages. It’s actually easy. Let’s give it a try by installing and
using the <code>limma</code> package to analyze some gene expression data.</p>
<p>If you’re recall from before, we can use <code>::</code> to avoid having to actually
<code>library</code> the whole package:</p>
<div class="sourceCode" id="cb450"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb450-1" title="1">BiocManager<span class="op">::</span><span class="kw">install</span>()</a></code></pre></div>
<p>And so we can use this to install important packages like <code>DESeq2</code>.</p>
<div class="sourceCode" id="cb451"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb451-1" title="1">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&#39;DESeq2&#39;</span>)</a></code></pre></div>
<p>Hold off on installing this quite yet. It’d take a while.
In Bioconductor, I always say to update ‘none’ (press ‘n’ when prompted).</p>
<p>There’s lots of good documentation on each
<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">package’s webpage</a>,
like <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">vingettes</a>.</p>
</div>
<div id="what-packages-do-you-have-already" class="section level3">
<h3><span class="header-section-number">4.3.5</span> What packages do you have already?</h3>
<p>You have a few. Here’s a list:</p>
<div class="sourceCode" id="cb452"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb452-1" title="1"><span class="kw">installed.packages</span>()</a></code></pre></div>
<p>But which are loaded?</p>
<div class="sourceCode" id="cb453"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb453-1" title="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<p>The packages are in a directory, the location is stored in this:</p>
<div class="sourceCode" id="cb454"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb454-1" title="1"><span class="kw">.libPaths</span>()</a></code></pre></div>
<p>I’m not running that one, as it’ll be different for each user.</p>
<hr />
</div>
<div id="major-problems-with-package-installations" class="section level3">
<h3><span class="header-section-number">4.3.6</span> Major problems with package installations</h3>
<p>Somethings can go wrong. You’ll want to pay attention to what is output
from the installation attempt. Look for errors. Scroll up to find warnings
and errors.</p>
<p>These can arise from multiple ways. Here’s a few ideas:</p>
<ul>
<li>the package name you put in is misspelled</li>
<li>the package is in another repo (CRAN, Bioconductor, etc)</li>
<li>your internet is down (this is more of a Stanford WiFi problem)</li>
<li>your R package library directory is not writable (you don’t have permission)</li>
<li>the package is not available for this version of R</li>
<li>the package relies on the system having programs or software or libraries
installed that are not installed (less of a problem with RStudio)</li>
<li>the package actually has a bug</li>
</ul>
<p>Let’s next talk about how to find help for problems.</p>
<hr />

</div>
</div>
<div class="footnotes">
<hr />
<ol start="7">
<li id="fn7"><p>I have no idea where this is, never heard of that country!<a href="packages-install-and-use.html#fnref7" class="footnote-back">↩</a></p></li>
<li id="fn8"><p>Or, I use the one-liner in bash of
<code>Rscript -e 'rmarkdown::render("notebook.Rmd")'</code> <a href="packages-install-and-use.html#fnref8" class="footnote-back">↩</a></p></li>
</ol>
</div>
<p style="margin-top:10em;text-align:center;font-size:small;">
    <a href="https://github.com/darachm/dll-r/blob/main/LICENSE.md">
        Licensed Creative Commons Attribution-NonCommercial-ShareAlike 4.0
    </a>
</p>
<p style="text-align:center;font-size:small;">
    <a href="https://github.com/rstudio/bookdown" target="blank">
        Website typeset with bookdown
    </a>
</p>

<script> 
// This is to look through, everything that's a floating lone double dash,
// delete it. This is left over for stuff that's trying to be pauses for
// the slides
var listz = document.getElementsByTagName("p");
for (var i=0; i < listz.length; i++) { 
    if (listz[i].innerText === "–") { listz[i].innerText = "" } 
} 
</script>
            </section>

          </div>
        </div>
      </div>
<a href="analyzing-viral-protein-sequences.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="getting-help-troubleshooting.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/darachm/dll-r/edit/main/550-packages.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["DLL 2021 R syllabus site.html"],
"toc": {
"collapse": "subsection"
},
"highlight": "zenburn"
});
});
</script>

</body>

</html>
